// powerful-react-table/dist/components/Filter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Filter;
var _react = _interopRequireWildcard(require("react"));
var _useTable = require("../useTable");
var _localeConfig = require("../../localeConfig");
var _setTableColumnWidths = _interopRequireDefault(require("../Table/setTableColumnWidths"));
var _useOut = _interopRequireDefault(require("../../useOut"));
var _Radio = _interopRequireDefault(require("../Radio"));
var _Input = _interopRequireDefault(require("../Input"));
var _Checkbox = _interopRequireDefault(require("../Checkbox"));
var _DatePicker = _interopRequireDefault(require("../DatePicker"));
var _pi = require("react-icons/pi");
var _md = require("react-icons/md");
var _hi = require("react-icons/hi");
var _fi = require("react-icons/fi");
require("./Filter.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function Filter(_ref) {
  let {
    column,
    headers,
    createRef
  } = _ref;
  const [displayFilterOptions, setDisplayFilterOptions] = (0, _react.useState)(false);
  const [radioOptions, setRadioOptions] = (0, _react.useState)([]);
  const [currentSelection, setCurrentSelection] = (0, _react.useState)('');
  const [selected, setSelected] = (0, _react.useState)([]);
  const [is, setIs] = (0, _react.useState)('');
  const [contains, setContains] = (0, _react.useState)('');
  const [between, setBetween] = (0, _react.useState)({
    min: 0,
    max: 0
  });
  const [dates, setDates] = (0, _react.useState)({
    start: null,
    end: null
  });
  const [dropdownAlignment, setDropdownAlignment] = (0, _react.useState)('left');
  const {
    data,
    selectedHeader,
    searchFilter,
    setSearchFilter,
    tableId,
    setOpenDropdown,
    setHiddenColumnIds
  } = (0, _useTable.useTable)();
  const containerRef = (0, _react.useRef)(null);
  const dropdownRef = (0, _react.useRef)(null);
  const localeText = (0, _localeConfig.getLocaleText)();
  (0, _react.useEffect)(() => {
    switch (column?.dataType) {
      case 'text':
        setRadioOptions([{
          value: 'contains',
          label: localeText.headerFilterOperatorContains,
          element: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Input.default, {
            className: "fc-option-list__input",
            focused: true,
            onChange: value => setContains(value)
          })
        }]);
        setCurrentSelection('contains');
        break;
      case 'option':
        setRadioOptions([{
          value: 'option',
          element: /*#__PURE__*/(0, _jsxRuntime.jsx)(Select, {
            options: getColumnData(selectedHeader),
            onChange: value => setSelected(value)
          })
        }]);
        setCurrentSelection('option');
        break;
      case 'between':
        setRadioOptions([{
          value: 'between',
          label: localeText.headerFilterOperatorBetween,
          element: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "d-flex align-items-center gap-1",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Input.default, {
              value: between?.min,
              focused: true,
              onChange: value => setBetween(prevBetween => ({
                ...prevBetween,
                min: value
              })),
              className: "fc-option-list__input-between"
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Input.default, {
              value: between?.max,
              onChange: value => setBetween(prevBetween => ({
                ...prevBetween,
                max: value
              })),
              className: "fc-option-list__input-between"
            })]
          })
        }]);
        setCurrentSelection('between');
        break;
      case 'date':
        setRadioOptions([{
          value: 'between',
          label: localeText.headerFilterOperatorBetween,
          element: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "d-flex align-items-center gap-2",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_DatePicker.default, {
              onChange: value => setDates(prevDates => ({
                ...prevDates,
                start: value
              }))
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_DatePicker.default, {
              onChange: value => setDates(prevDates => ({
                ...prevDates,
                end: value
              }))
            })]
          })
        }]);
        setCurrentSelection('between');
        break;
      default:
        break;
    }
  }, [column]);
  (0, _react.useEffect)(() => {
    calculatePosition();
  }, []);
  (0, _react.useEffect)(() => {
    window.addEventListener('resize', calculatePosition);
    return () => {
      window.removeEventListener('resize', calculatePosition);
    };
  }, []);
  (0, _useOut.default)(containerRef, () => setOpenDropdown(false));
  const handleDisplayOptions = () => {
    setDisplayFilterOptions(true);
  };
  const getColumnData = columnIndex => {
    const array = [];
    data?.forEach(element => {
      const childNodes = element?.props?.children;
      if (childNodes && childNodes[columnIndex]) {
        // Usa React.Children.toArray para manipular os children de forma mais segura
        const columnChild = _react.default.Children.toArray(childNodes[columnIndex].props.children);
        // Mapeia para obter os textos ou outros valores dentro dos elementos
        array.push(columnChild.map(child => {
          if (typeof child === 'string' || typeof child === 'number') {
            return child; // Retorna o texto ou número diretamente
          } else if (/*#__PURE__*/_react.default.isValidElement(child)) {
            return _react.default.Children.toArray(child.props.children).join(''); // Acessa o children interno
          }
          return null;
        }).join(''));
      }
    });
    const uniqueValues = [...new Set(array.filter(item => item !== null))];
    return uniqueValues.map(value => ({
      label: value,
      value: value,
      checked: false
    }));
  };
  const handleApplyFilter = () => {
    setOpenDropdown(false);
    if (selected?.length) {
      changeDispatch({
        type: 'option',
        value: selected?.map(_ref2 => {
          let {
            value
          } = _ref2;
          return value;
        })
      });
      return;
    }
    if (is || contains) {
      changeDispatch({
        type: 'text',
        value: is || contains
      });
      return;
    }
    if (between?.min && between?.max) {
      changeDispatch({
        type: 'between',
        value: between
      });
      return;
    }
    if (dates.start && dates.end) {
      changeDispatch({
        type: 'date',
        value: dates
      });
      return;
    }
  };
  const handleClearFilter = () => {
    const storage = localStorage.getItem('__magnunTablePreferences');
    const savedFilter = JSON.parse(storage)?.filters;
    if (!savedFilter) return setOpenDropdown(false);
    delete savedFilter[tableId][column?.name?.toUpperCase()];
    localStorage.setItem('__magnunTablePreferences', JSON.stringify({
      ...JSON.parse(storage),
      filters: savedFilter
    }));
    setSearchFilter(savedFilter);
    setOpenDropdown(false);
  };
  const changeDispatch = value => {
    const columnName = column?.name;
    const updatedSearchFilter = searchFilter.filter(filter => filter.label !== columnName);
    setSearchFilter([...updatedSearchFilter, {
      ...value,
      label: columnName,
      position: selectedHeader
    }]);
    const storage = localStorage.getItem('__magnunTablePreferences');
    const filtersStorage = JSON.parse(storage)?.filters || {};
    const newValues = {
      ...filtersStorage,
      [tableId]: {
        ...filtersStorage?.[tableId],
        [columnName?.toUpperCase()]: {
          label: columnName,
          position: selectedHeader,
          type: value?.type,
          value: value?.value
        }
      }
    };
    localStorage.setItem('__magnunTablePreferences', JSON.stringify({
      ...JSON.parse(storage),
      filters: newValues
    }));
  };
  const handleAutosize = () => {
    (0, _setTableColumnWidths.default)(data, headers, selectedHeader);
    setOpenDropdown(false);
  };
  const calculatePosition = () => {
    const widths = 255;
    if (dropdownRef.current && createRef) {
      const parentRect = createRef.getBoundingClientRect();
      const dropdownRect = dropdownRef.current.getBoundingClientRect();
      const spaceRight = window.innerWidth - parentRect.right;
      const spaceAbove = parentRect.top;
      const spaceBelow = window.innerHeight - parentRect.bottom;
      let newPosition = {
        left: spaceRight >= dropdownRect.width ? parentRect.left - 230 : parentRect.right - widths
      };
      const alignment = spaceRight >= dropdownRect.width ? 'left' : 'right';
      setDropdownAlignment(alignment);
      setTimeout(() => {
        const containerHeight = dropdownRef?.current?.getBoundingClientRect()?.height + 115;
        if (spaceAbove < spaceBelow) {
          newPosition.top = parentRect.bottom + 10;
        } else {
          newPosition.top = parentRect.top - containerHeight;
        }
        if (containerRef) {
          containerRef.current.style.left = `${newPosition.left}px`;
          containerRef.current.style.top = `${newPosition.top}px`;
        }
      }, 1);
    }
  };
  const handleHideColumn = () => {
    setHiddenColumnIds(prevState => prevState.includes(selectedHeader) ? prevState.filter(id => id !== selectedHeader) : [...prevState, selectedHeader]);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: `prt-filter-container ${displayFilterOptions && 'prt-filter-container--reset-border-top-right-radius'}`,
    ref: containerRef,
    style: {
      ...(dropdownAlignment == 'right' && {
        borderTopLeftRadius: 0
      })
    },
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("ul", {
      className: "prt-filter-container__primary-list",
      ref: dropdownRef,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("li", {
        className: "prt-filter-container__primary-list__item",
        onClick: handleDisplayOptions,
        children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "d-flex align-items-center justify-content-between w-100",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "d-flex align-items-center",
            style: {
              gap: '12px'
            },
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_pi.PiSlidersHorizontal, {
              style: {
                fontSize: '24px'
              }
            }), localeText.columnMenuFilterBy, " ", column?.name]
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_fi.FiChevronRight, {})]
        })
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("li", {
        className: "prt-filter-container__primary-list__item",
        onClick: handleAutosize,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdExpand, {
          style: {
            fontSize: '20px',
            transform: 'rotate(90deg)'
          }
        }), localeText.columnMenuAutosizeThisColumn]
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "prt-filter-container__horizontal-line"
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {
      className: "prt-filter-container__secondary-list",
      children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("li", {
        className: "prt-filter-container__primary-list__item",
        onClick: handleHideColumn,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_hi.HiOutlineEyeOff, {
          style: {
            fontSize: '20px'
          }
        }), localeText.columnMenuHideColumn]
      })
    }), displayFilterOptions && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: `fc-option-list ${dropdownAlignment == 'right' && 'fc-option-list__reverse'}`,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Radio.default, {
        options: radioOptions,
        selected: currentSelection
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: "fc-option-list__buttons",
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
          className: "fc-option-list__buttons__clear",
          onClick: handleClearFilter,
          children: localeText.filterBoxClearFilter
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
          className: "fc-option-list__buttons__apply",
          onClick: handleApplyFilter,
          children: localeText.filterBoxApplyFilter
        })]
      })]
    })]
  });
}
const Select = _ref3 => {
  let {
    options = [],
    onChange
  } = _ref3;
  const [values, setValues] = (0, _react.useState)(options);
  const handleSelection = (checked, index) => {
    values[index].checked = checked;
    setValues([...values]);
    onChange(values?.filter(value => value?.checked));
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "fc-option-list__select",
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Input.default, {
      placeholder: "Find option...",
      styles: {
        width: '219px'
      }
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "mt-4",
      children: values?.map((_ref4, index) => {
        let {
          label,
          value
        } = _ref4;
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(_Checkbox.default, {
          value: value,
          label: label,
          onChange: _ref5 => {
            let {
              checked
            } = _ref5;
            return handleSelection(checked, index);
          }
        }, index);
      })
    })]
  });
};
