// powerful-react-table/dist/components/Pagination/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Pagination;
var _react = _interopRequireWildcard(require("react"));
var _md = require("react-icons/md");
require("./Pagination.scss");
var _localeConfig = require("../../localeConfig");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function Pagination(_ref) {
  let {
    totalPerPage,
    onChangePage,
    totalItems,
    options = [],
    pageSizeOptions = []
  } = _ref;
  const localeText = (0, _localeConfig.getLocaleText)();
  const [paginationInfo, setPaginationInfo] = (0, _react.useState)({
    itemsPerPage: 10,
    currentPage: 1,
    offset: 0
  });
  const [config, setConfig] = (0, _react.useState)({
    skipPrevious: true,
    previous: true,
    next: true,
    skipNext: true,
    selectQuantity: true
  });
  const [itemOptionsPerPage, setItemOptionsPerPage] = (0, _react.useState)([{
    value: 10,
    label: '10'
  }, {
    value: 15,
    label: '15'
  }, {
    value: 20,
    label: '20'
  }, {
    value: 25,
    label: '25'
  }]);
  (0, _react.useEffect)(() => {
    onChangePage?.({
      pagination: paginationInfo
    });
  }, []);
  (0, _react.useEffect)(() => {
    setPaginationInfo({
      ...paginationInfo,
      itemsPerPage: totalPerPage
    });
    onChangePage?.({
      pagination: {
        ...paginationInfo,
        itemsPerPage: totalPerPage
      }
    });
  }, []);
  (0, _react.useEffect)(() => {
    setConfig({
      ...config,
      ...options
    });
  }, []);
  (0, _react.useEffect)(() => {
    if (pageSizeOptions?.length) {
      const values = pageSizeOptions?.map(option => ({
        label: option,
        value: option
      }));
      setItemOptionsPerPage([...values]);
      setPaginationInfo({
        ...paginationInfo,
        itemsPerPage: pageSizeOptions[0]
      });
    }
  }, [pageSizeOptions]);
  const changeItemsPerPage = value => {
    const pagination = {
      ...paginationInfo,
      itemsPerPage: Number(value)
    };
    onChangePage?.({
      pagination
    });
    setPaginationInfo({
      ...pagination
    });
  };
  const previous = page => {
    const previousPage = page > 1 ? page - 1 : 1;
    const pagination = {
      ...paginationInfo,
      currentPage: previousPage,
      offset: (previousPage - 1) * paginationInfo.itemsPerPage
    };
    onChangePage?.({
      pagination
    });
    setPaginationInfo({
      ...pagination
    });
  };
  const next = page => {
    const nextPage = page * paginationInfo.itemsPerPage < totalItems ? page + 1 : 1;
    const pagination = {
      ...paginationInfo,
      currentPage: nextPage,
      offset: (nextPage - 1) * paginationInfo.itemsPerPage
    };
    onChangePage?.({
      pagination
    });
    setPaginationInfo({
      ...pagination
    });
  };
  const firstOffset = () => {
    const pagination = {
      ...paginationInfo,
      currentPage: 1,
      offset: 0
    };
    onChangePage?.({
      pagination
    });
    setPaginationInfo({
      ...pagination
    });
  };
  const lastOffset = () => {
    const currentPage = Math.ceil(totalItems / paginationInfo.itemsPerPage);
    const pagination = {
      ...paginationInfo,
      currentPage,
      offset: (currentPage - 1) * paginationInfo.itemsPerPage
    };
    onChangePage?.({
      pagination
    });
    setPaginationInfo({
      ...pagination
    });
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: 'prt-pagination-container',
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: 'prt-pagination-container__controls',
      children: [config?.selectQuantity && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: 'quantity__label',
          children: localeText.rowsPerPageLabel
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: 'pagination__select',
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)("select", {
            name: "select-quantity",
            className: 'select__quantity',
            onChange: event => changeItemsPerPage(event.target.value),
            children: itemOptionsPerPage.map((option, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)("option", {
              value: option.value,
              children: option.label
            }, index))
          })
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: 'items__perpage',
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
          children: paginationInfo.currentPage === 1 ? totalItems > 0 ? 1 : 0 : paginationInfo.currentPage * paginationInfo.itemsPerPage - paginationInfo.itemsPerPage + 1
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
          children: "-"
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
          children: paginationInfo.currentPage * paginationInfo.itemsPerPage > totalItems ? totalItems : paginationInfo.currentPage * paginationInfo.itemsPerPage
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
          children: [" de ", totalItems]
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: 'navigation__buttons',
        children: [config?.skipPrevious && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: `action__button ${paginationInfo.currentPage === 1 ? 'disabled' : ''}`,
          onClick: () => firstOffset(),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineSkipPrevious, {})
        }), config?.previous && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: `action__button ${paginationInfo.currentPage === 1 ? 'disabled' : ''}`,
          onClick: () => previous(paginationInfo.currentPage),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdNavigateBefore, {})
        }), config?.next && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: `action__button ${paginationInfo.currentPage * paginationInfo.itemsPerPage >= totalItems ? 'disabled' : ''}`,
          onClick: () => next(paginationInfo.currentPage),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdNavigateNext, {})
        }), config?.skipNext && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: `action__button ${paginationInfo.currentPage * paginationInfo.itemsPerPage >= totalItems ? 'disabled' : ''}`,
          onClick: () => lastOffset(),
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineSkipNext, {})
        })]
      })]
    })
  });
}
