// react-view-files/dist/components/MediaPreview/index.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaPreview = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactPdf = require("react-pdf");
var _DropdownMenuOptions = _interopRequireDefault(require("../DropdownMenuOptions"));
var _Modal = require("../Modal");
var _Input = _interopRequireDefault(require("../Input"));
var _md = require("react-icons/md");
var _fa = require("react-icons/fa6");
var _fi = require("react-icons/fi");
var _extensions = require("../../extensions");
var _audioBox = _interopRequireDefault(require("../../img/audio-box.jpg"));
require("./MediaPreview.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return e; }; var t, e = {}, r = Object.prototype, n = r.hasOwnProperty, o = Object.defineProperty || function (t, e, r) { t[e] = r.value; }, i = "function" == typeof Symbol ? Symbol : {}, a = i.iterator || "@@iterator", c = i.asyncIterator || "@@asyncIterator", u = i.toStringTag || "@@toStringTag"; function define(t, e, r) { return Object.defineProperty(t, e, { value: r, enumerable: !0, configurable: !0, writable: !0 }), t[e]; } try { define({}, ""); } catch (t) { define = function define(t, e, r) { return t[e] = r; }; } function wrap(t, e, r, n) { var i = e && e.prototype instanceof Generator ? e : Generator, a = Object.create(i.prototype), c = new Context(n || []); return o(a, "_invoke", { value: makeInvokeMethod(t, r, c) }), a; } function tryCatch(t, e, r) { try { return { type: "normal", arg: t.call(e, r) }; } catch (t) { return { type: "throw", arg: t }; } } e.wrap = wrap; var h = "suspendedStart", l = "suspendedYield", f = "executing", s = "completed", y = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var p = {}; define(p, a, function () { return this; }); var d = Object.getPrototypeOf, v = d && d(d(values([]))); v && v !== r && n.call(v, a) && (p = v); var g = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(p); function defineIteratorMethods(t) { ["next", "throw", "return"].forEach(function (e) { define(t, e, function (t) { return this._invoke(e, t); }); }); } function AsyncIterator(t, e) { function invoke(r, o, i, a) { var c = tryCatch(t[r], t, o); if ("throw" !== c.type) { var u = c.arg, h = u.value; return h && "object" == _typeof(h) && n.call(h, "__await") ? e.resolve(h.__await).then(function (t) { invoke("next", t, i, a); }, function (t) { invoke("throw", t, i, a); }) : e.resolve(h).then(function (t) { u.value = t, i(u); }, function (t) { return invoke("throw", t, i, a); }); } a(c.arg); } var r; o(this, "_invoke", { value: function value(t, n) { function callInvokeWithMethodAndArg() { return new e(function (e, r) { invoke(t, n, e, r); }); } return r = r ? r.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); } }); } function makeInvokeMethod(e, r, n) { var o = h; return function (i, a) { if (o === f) throw new Error("Generator is already running"); if (o === s) { if ("throw" === i) throw a; return { value: t, done: !0 }; } for (n.method = i, n.arg = a;;) { var c = n.delegate; if (c) { var u = maybeInvokeDelegate(c, n); if (u) { if (u === y) continue; return u; } } if ("next" === n.method) n.sent = n._sent = n.arg;else if ("throw" === n.method) { if (o === h) throw o = s, n.arg; n.dispatchException(n.arg); } else "return" === n.method && n.abrupt("return", n.arg); o = f; var p = tryCatch(e, r, n); if ("normal" === p.type) { if (o = n.done ? s : l, p.arg === y) continue; return { value: p.arg, done: n.done }; } "throw" === p.type && (o = s, n.method = "throw", n.arg = p.arg); } }; } function maybeInvokeDelegate(e, r) { var n = r.method, o = e.iterator[n]; if (o === t) return r.delegate = null, "throw" === n && e.iterator["return"] && (r.method = "return", r.arg = t, maybeInvokeDelegate(e, r), "throw" === r.method) || "return" !== n && (r.method = "throw", r.arg = new TypeError("The iterator does not provide a '" + n + "' method")), y; var i = tryCatch(o, e.iterator, r.arg); if ("throw" === i.type) return r.method = "throw", r.arg = i.arg, r.delegate = null, y; var a = i.arg; return a ? a.done ? (r[e.resultName] = a.value, r.next = e.nextLoc, "return" !== r.method && (r.method = "next", r.arg = t), r.delegate = null, y) : a : (r.method = "throw", r.arg = new TypeError("iterator result is not an object"), r.delegate = null, y); } function pushTryEntry(t) { var e = { tryLoc: t[0] }; 1 in t && (e.catchLoc = t[1]), 2 in t && (e.finallyLoc = t[2], e.afterLoc = t[3]), this.tryEntries.push(e); } function resetTryEntry(t) { var e = t.completion || {}; e.type = "normal", delete e.arg, t.completion = e; } function Context(t) { this.tryEntries = [{ tryLoc: "root" }], t.forEach(pushTryEntry, this), this.reset(!0); } function values(e) { if (e || "" === e) { var r = e[a]; if (r) return r.call(e); if ("function" == typeof e.next) return e; if (!isNaN(e.length)) { var o = -1, i = function next() { for (; ++o < e.length;) if (n.call(e, o)) return next.value = e[o], next.done = !1, next; return next.value = t, next.done = !0, next; }; return i.next = i; } } throw new TypeError(_typeof(e) + " is not iterable"); } return GeneratorFunction.prototype = GeneratorFunctionPrototype, o(g, "constructor", { value: GeneratorFunctionPrototype, configurable: !0 }), o(GeneratorFunctionPrototype, "constructor", { value: GeneratorFunction, configurable: !0 }), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, u, "GeneratorFunction"), e.isGeneratorFunction = function (t) { var e = "function" == typeof t && t.constructor; return !!e && (e === GeneratorFunction || "GeneratorFunction" === (e.displayName || e.name)); }, e.mark = function (t) { return Object.setPrototypeOf ? Object.setPrototypeOf(t, GeneratorFunctionPrototype) : (t.__proto__ = GeneratorFunctionPrototype, define(t, u, "GeneratorFunction")), t.prototype = Object.create(g), t; }, e.awrap = function (t) { return { __await: t }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, c, function () { return this; }), e.AsyncIterator = AsyncIterator, e.async = function (t, r, n, o, i) { void 0 === i && (i = Promise); var a = new AsyncIterator(wrap(t, r, n, o), i); return e.isGeneratorFunction(r) ? a : a.next().then(function (t) { return t.done ? t.value : a.next(); }); }, defineIteratorMethods(g), define(g, u, "Generator"), define(g, a, function () { return this; }), define(g, "toString", function () { return "[object Generator]"; }), e.keys = function (t) { var e = Object(t), r = []; for (var n in e) r.push(n); return r.reverse(), function next() { for (; r.length;) { var t = r.pop(); if (t in e) return next.value = t, next.done = !1, next; } return next.done = !0, next; }; }, e.values = values, Context.prototype = { constructor: Context, reset: function reset(e) { if (this.prev = 0, this.next = 0, this.sent = this._sent = t, this.done = !1, this.delegate = null, this.method = "next", this.arg = t, this.tryEntries.forEach(resetTryEntry), !e) for (var r in this) "t" === r.charAt(0) && n.call(this, r) && !isNaN(+r.slice(1)) && (this[r] = t); }, stop: function stop() { this.done = !0; var t = this.tryEntries[0].completion; if ("throw" === t.type) throw t.arg; return this.rval; }, dispatchException: function dispatchException(e) { if (this.done) throw e; var r = this; function handle(n, o) { return a.type = "throw", a.arg = e, r.next = n, o && (r.method = "next", r.arg = t), !!o; } for (var o = this.tryEntries.length - 1; o >= 0; --o) { var i = this.tryEntries[o], a = i.completion; if ("root" === i.tryLoc) return handle("end"); if (i.tryLoc <= this.prev) { var c = n.call(i, "catchLoc"), u = n.call(i, "finallyLoc"); if (c && u) { if (this.prev < i.catchLoc) return handle(i.catchLoc, !0); if (this.prev < i.finallyLoc) return handle(i.finallyLoc); } else if (c) { if (this.prev < i.catchLoc) return handle(i.catchLoc, !0); } else { if (!u) throw new Error("try statement without catch or finally"); if (this.prev < i.finallyLoc) return handle(i.finallyLoc); } } } }, abrupt: function abrupt(t, e) { for (var r = this.tryEntries.length - 1; r >= 0; --r) { var o = this.tryEntries[r]; if (o.tryLoc <= this.prev && n.call(o, "finallyLoc") && this.prev < o.finallyLoc) { var i = o; break; } } i && ("break" === t || "continue" === t) && i.tryLoc <= e && e <= i.finallyLoc && (i = null); var a = i ? i.completion : {}; return a.type = t, a.arg = e, i ? (this.method = "next", this.next = i.finallyLoc, y) : this.complete(a); }, complete: function complete(t, e) { if ("throw" === t.type) throw t.arg; return "break" === t.type || "continue" === t.type ? this.next = t.arg : "return" === t.type ? (this.rval = this.arg = t.arg, this.method = "return", this.next = "end") : "normal" === t.type && e && (this.next = e), y; }, finish: function finish(t) { for (var e = this.tryEntries.length - 1; e >= 0; --e) { var r = this.tryEntries[e]; if (r.finallyLoc === t) return this.complete(r.completion, r.afterLoc), resetTryEntry(r), y; } }, "catch": function _catch(t) { for (var e = this.tryEntries.length - 1; e >= 0; --e) { var r = this.tryEntries[e]; if (r.tryLoc === t) { var n = r.completion; if ("throw" === n.type) { var o = n.arg; resetTryEntry(r); } return o; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(e, r, n) { return this.delegate = { iterator: values(e), resultName: r, nextLoc: n }, "next" === this.method && (this.arg = t), y; } }, e; }
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }
function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
var docx = require("docx-preview");
var html2canvas = require("html2canvas");
var MediaPreview = exports.MediaPreview = function MediaPreview(_ref) {
  var _ref$file = _ref.file,
    file = _ref$file === void 0 ? {} : _ref$file,
    _ref$displayFooter = _ref.displayFooter,
    displayFooter = _ref$displayFooter === void 0 ? true : _ref$displayFooter,
    _ref$hideDropdown = _ref.hideDropdown,
    hideDropdown = _ref$hideDropdown === void 0 ? false : _ref$hideDropdown,
    _ref$style = _ref.style,
    style = _ref$style === void 0 ? {} : _ref$style,
    _ref$cover = _ref.cover,
    cover = _ref$cover === void 0 ? false : _ref$cover,
    _ref$className = _ref.className,
    className = _ref$className === void 0 ? '' : _ref$className,
    onClick = _ref.onClick,
    onRename = _ref.onRename,
    onDelete = _ref.onDelete;
  var _useState = (0, _react.useState)(null),
    _useState2 = _slicedToArray(_useState, 2),
    selectedFile = _useState2[0],
    setSelectedFile = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    thumbnail = _useState4[0],
    setThumbnail = _useState4[1];
  var _useState5 = (0, _react.useState)(null),
    _useState6 = _slicedToArray(_useState5, 2),
    fileExtension = _useState6[0],
    setFileExtension = _useState6[1];
  var _useState7 = (0, _react.useState)(false),
    _useState8 = _slicedToArray(_useState7, 2),
    activeRename = _useState8[0],
    setActiveRename = _useState8[1];
  var _useState9 = (0, _react.useState)(""),
    _useState10 = _slicedToArray(_useState9, 2),
    changeFilename = _useState10[0],
    setChangeFilename = _useState10[1];
  var _useState11 = (0, _react.useState)("100%"),
    _useState12 = _slicedToArray(_useState11, 2),
    containerWidth = _useState12[0],
    setContainerWidth = _useState12[1];
  var _useState13 = (0, _react.useState)("0 K"),
    _useState14 = _slicedToArray(_useState13, 2),
    fileSize = _useState14[0],
    setFileSize = _useState14[1];
  var mediaPreviewRef = (0, _react.useRef)(null);
  var containerRef = (0, _react.useRef)(null);
  var loadRef = (0, _react.useRef)(null);
  var translate = function translate(text) {
    return text;
  };
  var icons = {
    image: require("../../img/picture.png"),
    excel: require("../../img/excel.png"),
    audio: require("../../img/audio.png"),
    video: require("../../img/play.png"),
    word: require("../../img/word.png"),
    pdf: require("../../img/pdf.png"),
    txt: require("../../img/news.png")
  };
  var dropdownOptions = [{
    name: translate("Visualizar"),
    icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineRemoveRedEye, {}),
    redirect: false,
    action: "view"
  }, {
    name: translate("Renomear"),
    icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineDriveFileRenameOutline, {}),
    redirect: false,
    action: "rename"
  }, {
    name: translate("Download"),
    icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineFileDownload, {}),
    redirect: false,
    action: "download"
  }, {
    name: translate("Remover"),
    icon: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdDeleteOutline, {}),
    redirect: false,
    action: "delete"
  }];
  var backgroundSize = {
    pdf: '220px',
    word: '480px',
    txt: '220px'
  };
  var backgroundPosition = {
    pdf: 'center',
    word: 'center -30px',
    txt: 'center 10px'
  };
  var word = _extensions.extensionGroup.word,
    pdf = _extensions.extensionGroup.pdf,
    image = _extensions.extensionGroup.image,
    video = _extensions.extensionGroup.video,
    audio = _extensions.extensionGroup.audio,
    excel = _extensions.extensionGroup.excel,
    txt = _extensions.extensionGroup.txt;
  var allExtensions = [].concat(_toConsumableArray(word), _toConsumableArray(pdf), _toConsumableArray(image), _toConsumableArray(video), _toConsumableArray(audio), _toConsumableArray(txt));
  (0, _react.useEffect)(function () {
    setSelectedFile((file === null || file === void 0 ? void 0 : file.file) || file);
    loadExtensions((file === null || file === void 0 ? void 0 : file.file) || file);
    generateFileThumbnail((file === null || file === void 0 ? void 0 : file.file) || file);
    if (cover) getFileSize((file === null || file === void 0 ? void 0 : file.file) || file);
  }, [file]);
  (0, _react.useEffect)(function () {
    var _mediaPreviewRef$curr;
    if (mediaPreviewRef !== null && mediaPreviewRef !== void 0 && (_mediaPreviewRef$curr = mediaPreviewRef.current) !== null && _mediaPreviewRef$curr !== void 0 && _mediaPreviewRef$curr.offsetWidth) {
      setContainerWidth("".concat(mediaPreviewRef.current.offsetWidth, "px"));
    }
  }, [thumbnail]);
  (0, _react.useEffect)(function () {
    window.addEventListener("resize", function () {
      var _mediaPreviewRef$curr2;
      if (mediaPreviewRef !== null && mediaPreviewRef !== void 0 && (_mediaPreviewRef$curr2 = mediaPreviewRef.current) !== null && _mediaPreviewRef$curr2 !== void 0 && _mediaPreviewRef$curr2.offsetWidth) {
        setContainerWidth("".concat(mediaPreviewRef.current.offsetWidth, "px"));
      }
    });
  }, []);
  var loadExtensions = function loadExtensions(item) {
    if (word.includes(item === null || item === void 0 ? void 0 : item.type)) setFileExtension({
      type: "word",
      name: _extensions.extensions[item === null || item === void 0 ? void 0 : item.type]
    });
    if (pdf.includes(item === null || item === void 0 ? void 0 : item.type)) setFileExtension({
      type: "pdf",
      name: _extensions.extensions[item === null || item === void 0 ? void 0 : item.type]
    });
    if (image.includes(item === null || item === void 0 ? void 0 : item.type)) setFileExtension({
      type: "image",
      name: _extensions.extensions[item === null || item === void 0 ? void 0 : item.type]
    });
    if (video.includes(item === null || item === void 0 ? void 0 : item.type)) setFileExtension({
      type: "video",
      name: _extensions.extensions[item === null || item === void 0 ? void 0 : item.type]
    });
    if (audio.includes(item === null || item === void 0 ? void 0 : item.type)) setFileExtension({
      type: "audio",
      name: _extensions.extensions[item === null || item === void 0 ? void 0 : item.type]
    });
    if (excel.includes(item === null || item === void 0 ? void 0 : item.type)) setFileExtension({
      type: "excel",
      name: _extensions.extensions[item === null || item === void 0 ? void 0 : item.type]
    });
    if (txt.includes(item === null || item === void 0 ? void 0 : item.type)) setFileExtension({
      type: "txt",
      name: _extensions.extensions[item === null || item === void 0 ? void 0 : item.type]
    });
  };
  var generateFileThumbnail = function generateFileThumbnail(item) {
    if (image.includes(item === null || item === void 0 ? void 0 : item.type)) setThumbnail((item === null || item === void 0 ? void 0 : item.url) || URL.createObjectURL(item));
    if (audio.includes(item === null || item === void 0 ? void 0 : item.type)) setThumbnail(_audioBox["default"]);
    if (pdf.includes(item === null || item === void 0 ? void 0 : item.type)) generatePDFThumbnail(item);
    if (word.includes(item === null || item === void 0 ? void 0 : item.type)) generateWordThumbnail(item);
    if (video.includes(item === null || item === void 0 ? void 0 : item.type)) {
      handleFileUpload(item);
    }
    if (txt.includes(item === null || item === void 0 ? void 0 : item.type)) generateTXTThumbnail(item);
  };
  var handleFileUpload = /*#__PURE__*/function () {
    var _ref2 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(file) {
      var thumbnail;
      return _regeneratorRuntime().wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return generateVideoThumbnail(file, 1.55);
          case 2:
            thumbnail = _context.sent;
            setThumbnail(thumbnail);
          case 4:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function handleFileUpload(_x) {
      return _ref2.apply(this, arguments);
    };
  }();
  var generateVideoThumbnail = function generateVideoThumbnail(file, timeInSeconds) {
    return new Promise(function (resolve) {
      var canvas = document.createElement("canvas");
      var video = document.createElement("video");
      video.autoplay = true;
      video.muted = true;
      video.src = file !== null && file !== void 0 && file.url ? file.url : URL.createObjectURL(file);
      video.onloadedmetadata = function () {
        video.currentTime = timeInSeconds; // Defina o tempo desejado aqui
      };

      video.onseeked = function () {
        // Cria o contexto do canvas e desenha o frame atual do vídeo
        canvas.width = video.videoWidth;
        canvas.height = video.videoHeight;
        var ctx = canvas.getContext("2d");
        ctx.drawImage(video, 0, 0, video.videoWidth, video.videoHeight);
        resolve(canvas.toDataURL("image/png"));
        video.pause();
      };
      video.onerror = function () {
        resolve(null); // Se houver um erro, retorna nulo
      };
    });
  };

  var generatePDFThumbnail = function generatePDFThumbnail(item) {
    var fileReader = new FileReader();
    fileReader.onload = function () {
      var pdfData = new Uint8Array(this.result);
      var loadingTask = _reactPdf.pdfjs.getDocument({
        data: pdfData
      });
      var pageNumber = 1;
      loadingTask.promise.then(function (pdf) {
        pdf.getPage(pageNumber).then(function (page) {
          var scale = 1.5;
          var viewport = page.getViewport({
            scale: scale
          });
          var canvas = document.createElement("canvas");
          var context = canvas.getContext("2d");
          canvas.height = viewport.height;
          canvas.width = viewport.width;
          var renderTask = page.render({
            canvasContext: context,
            viewport: viewport
          });
          renderTask.promise.then(function () {
            return setThumbnail(canvas.toDataURL());
          });
        });
      });
    };
    if (item instanceof File) {
      fileReader.readAsArrayBuffer(item);
      return;
    }
    fetch(item === null || item === void 0 ? void 0 : item.url).then(function (response) {
      return response.blob();
    }).then(function (blob) {
      return fileReader.readAsArrayBuffer(blob);
    });
  };
  var generateWordThumbnail = function generateWordThumbnail(item) {
    var options = {
      allowTaint: true,
      logging: true,
      useCORS: true,
      scrollY: 0,
      scrollX: 0,
      scale: 0.5
    };
    var element = document.createElement("div");
    element.setAttribute("id", "preview-docx");
    document.body.appendChild(element);
    if (item instanceof File) {
      docx.renderAsync(item, element).then(function () {
        html2canvas(element, options).then(function (canvas) {
          return setThumbnail(canvas.toDataURL("image/jpeg"));
        });
        element.remove();
      });
      return;
    }
    createFile(item).then(function (response) {
      docx.renderAsync(response, element).then(function () {
        html2canvas(element, options).then(function (canvas) {
          return setThumbnail(canvas.toDataURL("image/jpeg"));
        });
        element.remove();
      });
    });
  };
  var generateTXTThumbnail = /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(item) {
      var container, element, fileContent, response, lines, thumbnailWidth, thumbnailHeight, canvas;
      return _regeneratorRuntime().wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            container = document.createElement("div");
            container.style.position = "absolute";
            container.style.left = "-9999px";
            document.body.appendChild(container);
            element = document.createElement("div");
            element.setAttribute("id", "preview-txt");
            container.appendChild(element);
            _context2.prev = 7;
            if (!(item instanceof File)) {
              _context2.next = 14;
              break;
            }
            _context2.next = 11;
            return renderFileContent(element, item);
          case 11:
            fileContent = _context2.sent;
            _context2.next = 20;
            break;
          case 14:
            _context2.next = 16;
            return createFile(item);
          case 16:
            response = _context2.sent;
            _context2.next = 19;
            return renderFileContent(element, response);
          case 19:
            fileContent = _context2.sent;
          case 20:
            lines = breakTextIntoLines(fileContent, 280);
            element.style.height = "".concat(lines.length * 1.2, "em");
            element.innerText = lines.join("\n");
            thumbnailWidth = 280;
            thumbnailHeight = 280;
            _context2.next = 27;
            return html2canvas(element, {
              width: thumbnailWidth,
              height: thumbnailHeight
            });
          case 27:
            canvas = _context2.sent;
            setThumbnail(canvas.toDataURL("image/jpeg"));
          case 29:
            _context2.prev = 29;
            document.body.removeChild(container);
            return _context2.finish(29);
          case 32:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[7,, 29, 32]]);
    }));
    return function generateTXTThumbnail(_x2) {
      return _ref3.apply(this, arguments);
    };
  }();
  var renderFileContent = function renderFileContent(element, file) {
    return new Promise(function (resolve, reject) {
      var reader = new FileReader();
      reader.onload = function (e) {
        var fileContent = e.target.result;
        resolve(fileContent);
      };
      reader.onerror = function (error) {
        reject(error);
      };
      reader.readAsText(file);
    });
  };
  var breakTextIntoLines = function breakTextIntoLines(text, maxWidth) {
    var words = text === null || text === void 0 ? void 0 : text.split(" ");
    var currentLine = "";
    var lines = [];
    words.forEach(function (word) {
      var testLine = currentLine.length === 0 ? word : "".concat(currentLine, " ").concat(word);
      var testWidth = measureTextWidth(testLine);
      if (testWidth > maxWidth) {
        lines.push(currentLine.trim());
        currentLine = word;
      } else {
        currentLine = testLine;
      }
    });
    lines.push(currentLine.trim());
    return lines;
  };
  var measureTextWidth = function measureTextWidth(text) {
    var canvas = document.createElement("canvas");
    var context = canvas.getContext("2d");
    context.font = getComputedStyle(document.body).font;
    return context.measureText(text).width;
  };
  function createFile(_x3) {
    return _createFile.apply(this, arguments);
  }
  function _createFile() {
    _createFile = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee3(item) {
      var response, data, metadata, file;
      return _regeneratorRuntime().wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return fetch(item.url);
          case 2:
            response = _context3.sent;
            _context3.next = 5;
            return response.blob();
          case 5:
            data = _context3.sent;
            metadata = {
              type: "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
            };
            file = new File([data], item.name, metadata);
            return _context3.abrupt("return", file);
          case 9:
          case "end":
            return _context3.stop();
        }
      }, _callee3);
    }));
    return _createFile.apply(this, arguments);
  }
  var handleDropdownSelection = function handleDropdownSelection(_ref4) {
    var action = _ref4.action;
    if (action === "view") onClick === null || onClick === void 0 || onClick(selectedFile, fileExtension);
    if (action === "download") downloadFile();
    if (action === "rename") setActiveRename(true);
    if (action === "delete") {
      onDelete === null || onDelete === void 0 || onDelete(file);
    }
  };
  var downloadFile = function downloadFile() {
    var url = selectedFile !== null && selectedFile !== void 0 && selectedFile.url ? selectedFile.url : window.URL.createObjectURL(selectedFile);
    var link = document.createElement("a");
    link.download = selectedFile.name;
    link.href = url;
    link.click();
    window.URL.revokeObjectURL(url);
  };
  var getFileSize = function getFileSize(item) {
    fetch(item === null || item === void 0 ? void 0 : item.url).then(function (response) {
      var _response$headers;
      var contentLength = response === null || response === void 0 || (_response$headers = response.headers) === null || _response$headers === void 0 ? void 0 : _response$headers.get("Content-Length");
      setFileSize(formatBytes(parseInt(contentLength, 10)));
    });
  };
  var formatBytes = function formatBytes(bytes) {
    var decimals = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 1;
    if (!bytes) return "0 Bytes";
    var k = 1024;
    var dm = decimals < 0 ? 0 : decimals;
    var sizes = ["B", "KB", "MB", "GB", "TB"];
    var i = Math.floor(Math.log(bytes) / Math.log(k));
    return "".concat(parseFloat((bytes / Math.pow(k, i)).toFixed(dm)), " ").concat(sizes[i]);
  };
  var handleRename = function handleRename() {
    setSelectedFile(_objectSpread(_objectSpread({}, selectedFile), {}, {
      name: changeFilename
    }));
    setActiveRename(false);
    onRename === null || onRename === void 0 || onRename(_objectSpread(_objectSpread({}, selectedFile), {}, {
      name: changeFilename
    }));
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "media-preview-container ".concat(!displayFooter && 'media-preview-container__hide-footer', " ").concat(className, " "),
      onDoubleClick: function onDoubleClick() {
        return onClick === null || onClick === void 0 ? void 0 : onClick(selectedFile, fileExtension);
      },
      ref: containerRef,
      children: [(thumbnail || !(allExtensions !== null && allExtensions !== void 0 && allExtensions.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type))) && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          id: "viewer-thumbnail",
          className: "viewer-thumbnail",
          style: {
            backgroundImage: "url(".concat(thumbnail, ")"),
            width: (style === null || style === void 0 ? void 0 : style.width) || "100%",
            height: (style === null || style === void 0 ? void 0 : style.height) || "170px",
            backgroundRepeat: "no-repeat",
            backgroundSize: (style === null || style === void 0 ? void 0 : style.backgroundSize) || backgroundSize[fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type] || 'cover',
            backgroundPosition: (style === null || style === void 0 ? void 0 : style.backgroundPosition) || backgroundPosition[fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type]
          },
          ref: mediaPreviewRef,
          children: (fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type) === 'video' && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "video-play-icon",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaPlay, {})
          })
        }), displayFooter && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "viewer-thumbnail__footer",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "d-flex align-items-center",
            style: {
              overflow: "hidden"
            },
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
              className: "file-icon",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
                src: icons[fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type],
                alt: selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.name
              })
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
              className: "d-flex align-items-center ml-2",
              style: {
                overflow: "hidden"
              },
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)("p", {
                style: {
                  textOverflow: "ellipsis",
                  overflow: "hidden",
                  whiteSpace: "nowrap"
                },
                title: selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.name,
                children: selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.name
              })
            })]
          }), !hideDropdown && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_DropdownMenuOptions["default"], {
              options: dropdownOptions,
              onClicked: function onClicked(value) {
                return handleDropdownSelection(value);
              }
            })
          })]
        }), cover && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "cover-preview",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "cover-preview__name",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
              className: "file-icon",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
                src: icons[fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type],
                alt: selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.name
              })
            }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
              children: [selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.name, ".", fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.name]
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "file-size",
            children: (selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.size) || fileSize
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "w-100 d-flex justify-content-center",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
              className: "file-download",
              onClick: function onClick() {
                return downloadFile();
              },
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fi.FiDownload, {})
            })
          })]
        })]
      }), !thumbnail && (allExtensions === null || allExtensions === void 0 ? void 0 : allExtensions.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type)) && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: "loading-preview",
        ref: loadRef,
        style: {
          minWidthidth: (style === null || style === void 0 ? void 0 : style.width) || "100%",
          width: (style === null || style === void 0 ? void 0 : style.width) || "100%",
          minHeight: (style === null || style === void 0 ? void 0 : style.height) || "225px",
          height: (style === null || style === void 0 ? void 0 : style.height) || "225px"
        },
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "loader"
        })
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Modal.Modal, {
      show: activeRename,
      alignment: "center",
      content: "small",
      children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_Modal.ModalBody, {
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
          className: "mb-3",
          children: "Renomear"
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Input["default"], {
          type: "text",
          className: "mpc-container__input-rename",
          value: selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.name,
          onFocused: true,
          onChange: function onChange(value) {
            return setChangeFilename(value);
          }
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "d-flex align-items-center justify-content-end gap-1 mt-3",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
            onClick: function onClick() {
              return setActiveRename(false);
            },
            className: "media-preview-button media-preview-button__cancel",
            children: "Cancelar"
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
            onClick: handleRename,
            className: "media-preview-button media-preview-button__save",
            children: "Salvar"
          })]
        })]
      })
    })]
  });
};
