// blend-editor/dist/components/Editor/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Editor;
var _react = _interopRequireWildcard(require("react"));
var _Dropdown = require("../Dropdown");
var _fa = require("react-icons/fa");
var _fa2 = require("react-icons/fa6");
require("./style.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function Editor(_ref) {
  var _ref2, _ref3, _ref4, _ref5, _ref6, _ref7;
  let {
    content,
    focused,
    onChange
  } = _ref;
  const [styles, setStyles] = (0, _react.useState)({
    bold: false,
    italic: false,
    underline: false,
    strikethrough: false,
    quote: false,
    listOl: false,
    listUl: false,
    code: false,
    h1: false,
    h2: false,
    h3: false,
    h4: false,
    h5: false,
    h6: false
  });
  const [hasContentRight, setHasContentRight] = (0, _react.useState)(false);
  const [temTextoSelecionado, setTemTextoSelecionado] = (0, _react.useState)(false);
  const [activeStyles, setActiveStyles] = (0, _react.useState)([]);
  const [textColor, setTextColor] = (0, _react.useState)('#162949');
  const [linkInfo, setLinkInfo] = (0, _react.useState)({
    url: '',
    title: ''
  });
  const [savedRange, setSavedRange] = (0, _react.useState)(null);
  const [userPressedEnter, setUserPressedEnter] = (0, _react.useState)(false);
  const [fontSelecionada, setFontSelecionada] = (0, _react.useState)('Texto normal');
  const [screenSize, setScreenSize] = (0, _react.useState)(window.innerWidth);
  const editorRef = (0, _react.useRef)(null);
  const createLinkRef = (0, _react.useRef)(null);
  const values = {
    bold: 'strong',
    italic: 'em',
    underline: 'u',
    strikethrough: 's',
    quote: 'blockquote',
    listOl: 'ol',
    listUl: 'ul',
    code: 'code',
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6'
  };
  const colors = ["#162949", "#0646A5", "#BE2500", "#FF8B01", "#016745", "#5D6B83", "#005FED", "#CA371C", "#F5CD47", "#32A877"];
  const opcoesDeFontes = [{
    value: 'p',
    label: 'Texto normal'
  }, {
    value: 'h1',
    label: 'Título 1'
  }, {
    value: 'h2',
    label: 'Título 2'
  }, {
    value: 'h3',
    label: 'Título 3'
  }, {
    value: 'h4',
    label: 'Título 4'
  }, {
    value: 'h5',
    label: 'Título 5'
  }, {
    value: 'h6',
    label: 'Título 6'
  }];
  var currentNode = null;
  var currentElement = null;
  (0, _react.useEffect)(() => {
    const container = document.getElementById('b-editor-container');
    if (container) {
      setScreenSize(parseInt(container.offsetWidth));
    }
  }, []);
  (0, _react.useEffect)(() => {
    if (editorRef.current) editorRef.current.innerHTML = content;
  }, [content]);
  (0, _react.useEffect)(() => {
    const lastChild = editorRef.current.lastChild;
    if (focused && lastChild) {
      editorRef.current.focus();
      const range = document.createRange();
      const sel = window.getSelection();
      range.setStartAfter(lastChild);
      range.collapse(true);
      sel.removeAllRanges();
      sel.addRange(range);
    }
  }, [focused]);
  (0, _react.useEffect)(() => {
    document.addEventListener('selectionchange', function () {
      var selection = window.getSelection();
      if (selection.toString().length > 0) {
        setTemTextoSelecionado(true);
      } else {
        setTemTextoSelecionado(false);
      }
    });
  });
  (0, _react.useEffect)(() => {
    setActiveStyles(Object.keys(styles).filter(style => styles[style]));
  }, [styles]);
  (0, _react.useEffect)(() => {
    const handleResize = () => {
      const container = document.getElementById('b-editor-container');
      if (container) {
        setScreenSize(parseInt(container.offsetWidth));
      }
    };
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);
  const handleKeyPress = event => {
    const char = String.fromCharCode(event.which);
    if (event.key === 'Enter') {
      setUserPressedEnter(true);
      resetStyles();
      setFontSelecionada('Texto normal');
      if (activeStyles.includes('code')) {
        event.preventDefault();
        const {
          element
        } = isCursorInsideEmOrStrong();
        let span = document.createElement('span');
        span.innerHTML = ' ';
        element.appendChild(span);
        const range = document.createRange();
        range.setStart(span.firstChild, 0);
        range.setEnd(span.firstChild, 0);
        const sel = window.getSelection();
        sel.removeAllRanges();
        sel.addRange(range);
      }
      if (!(styles.listOl || styles.listUl || styles.quote || styles.code) || !(activeStyles !== null && activeStyles !== void 0 && activeStyles.length)) insertParagraphWithBreak(event);
      return;
    }
    setUserPressedEnter(false);
    if (event.key === ' ' || event.key === 'Enter') {
      insertText(char);
    } else {
      insertText(char);
      event.preventDefault();
    }
    onChange === null || onChange === void 0 || onChange({
      value: editorRef.current.innerHTML
    });
  };
  const toggleButton = (type, button) => {
    const buttonAction = document.getElementById(button);
    if (temTextoSelecionado) {
      toggleStyle(values[type]);
      return;
    }
    if (buttonAction && buttonAction.classList.contains('active')) {
      buttonAction.classList.remove('active');
    } else {
      if (buttonAction) buttonAction.classList.add('active');
    }
    styles[type] = !styles[type];
    setActiveStyles(Object.keys(styles).filter(style => styles[style]));
    if (styles[type] && !hasContentRight) {
      let element;
      if (type === 'listOl' || type === 'listUl') {
        element = document.createElement(values[type]);
        const li = document.createElement('li');
        element.appendChild(li);
      } else if (type === 'quote') {
        element = document.createElement(values[type]);
        const p = document.createElement('p');
        p.textContent = '\u2060';
        element.appendChild(p);
      } else {
        element = document.createElement(values[type]);
      }
      editorRef.current.appendChild(element);
      const range = document.createRange();
      const sel = window.getSelection();
      range.setStart(element, 0);
      range.setEnd(element, 0);
      sel.removeAllRanges();
      sel.addRange(range);
    }
    editorRef.current.focus();
  };
  const selecionaFonteDoTexto = font => {
    setFontSelecionada(font.label);
    const values = styles;
    values[font.value] = !values[font.value];
    setStyles({
      ...values
    });
    setActiveStyles(Object.keys(values).filter(style => values[style]));
  };
  function toggleStyle(style) {
    const selection = window.getSelection();
    if (selection.rangeCount > 0) {
      const range = selection.getRangeAt(0);
      const ancestorNodes = [];
      let currentNode = range.commonAncestorContainer;
      while (currentNode) {
        if (currentNode.nodeType === Node.ELEMENT_NODE) {
          ancestorNodes.push(currentNode);
        }
        currentNode = currentNode.parentNode;
      }
      const filtered = ancestorNodes.filter(node => node.nodeName.toLowerCase() === style.toLowerCase());
      if (filtered.length > 0 && ancestorNodes.length > 0) {
        for (const node of ancestorNodes) {
          if (node.nodeName === filtered[0].nodeName) {
            if (filtered[0].childNodes) {
              const content = filtered[0].innerHTML;
              const parent = filtered[0].parentNode;
              parent.insertBefore(document.createTextNode(content), filtered[0]);
              parent.removeChild(filtered[0]);
            }
          }
        }
      } else {
        const element = document.createElement(style);
        element.appendChild(range.extractContents());
        range.deleteContents();
        range.insertNode(element);
        const newRange = document.createRange();
        newRange.selectNodeContents(element);
        selection.removeAllRanges();
        selection.addRange(newRange);
      }
    }
  }
  function rgbToHex(rgb) {
    if (!rgb) return;
    let [r, g, b] = rgb.match(/\d+/g);
    r = parseInt(r).toString(16);
    g = parseInt(g).toString(16);
    b = parseInt(b).toString(16);
    if (r.length == 1) r = "0" + r;
    if (g.length == 1) g = "0" + g;
    if (b.length == 1) b = "0" + b;
    return "#" + r + g + b;
  }
  function insertText(text) {
    var selection = window.getSelection();
    var range = selection.getRangeAt(0);
    currentElement = editorRef.current;
    var listaDeParagrafos = currentElement.querySelectorAll('p');
    const divsWithBr = editorRef.current.querySelectorAll('div');
    if ((divsWithBr === null || divsWithBr === void 0 ? void 0 : divsWithBr.length) > 0) {
      divsWithBr.forEach(divElement => {
        var _divElement$querySele;
        if (((_divElement$querySele = divElement.querySelectorAll('br')) === null || _divElement$querySele === void 0 ? void 0 : _divElement$querySele.length) === 1) {
          divElement.remove();
        }
      });
    }
    if (!(activeStyles !== null && activeStyles !== void 0 && activeStyles.length) && !hasContentRight) {
      if ((listaDeParagrafos === null || listaDeParagrafos === void 0 ? void 0 : listaDeParagrafos.length) > 0) {
        var _paragraph$firstChild;
        const paragraph = listaDeParagrafos[(listaDeParagrafos === null || listaDeParagrafos === void 0 ? void 0 : listaDeParagrafos.length) - 1];
        if ((paragraph === null || paragraph === void 0 || (_paragraph$firstChild = paragraph.firstChild) === null || _paragraph$firstChild === void 0 ? void 0 : _paragraph$firstChild.nodeName) === 'BR') {
          paragraph.firstChild.remove();
          currentElement.appendChild(paragraph);
        }
        currentElement = paragraph;
      } else {
        const paragraph = document.createElement('p');
        currentElement.appendChild(paragraph);
        currentElement = paragraph;
      }
    }
    if (activeStyles !== null && activeStyles !== void 0 && activeStyles.length && !(activeStyles !== null && activeStyles !== void 0 && activeStyles.includes('quote')) && !(activeStyles !== null && activeStyles !== void 0 && activeStyles.includes('listOl')) && !(activeStyles !== null && activeStyles !== void 0 && activeStyles.includes('listUl')) && !(activeStyles !== null && activeStyles !== void 0 && activeStyles.includes('code'))) {
      var _activeStyles$filter;
      if ((listaDeParagrafos === null || listaDeParagrafos === void 0 ? void 0 : listaDeParagrafos.length) > 0) {
        var _paragraph$firstChild2;
        const paragraph = listaDeParagrafos[(listaDeParagrafos === null || listaDeParagrafos === void 0 ? void 0 : listaDeParagrafos.length) - 1];
        if ((paragraph === null || paragraph === void 0 || (_paragraph$firstChild2 = paragraph.firstChild) === null || _paragraph$firstChild2 === void 0 ? void 0 : _paragraph$firstChild2.nodeName) === 'BR') {
          paragraph.firstChild.remove();
          currentElement.appendChild(paragraph);
        }
        currentElement = paragraph;
      } else {
        const paragraph = document.createElement('p');
        currentElement.appendChild(paragraph);
        currentElement = paragraph;
        editorRef.current.appendChild(currentElement);
      }
      activeStyles === null || activeStyles === void 0 || (_activeStyles$filter = activeStyles.filter(style => !['quote', 'listOl', 'listUl', 'code.'].includes(style))) === null || _activeStyles$filter === void 0 || _activeStyles$filter.forEach(style => {
        var _lastChild$tagName;
        let lastChild = currentElement.lastChild;
        if (lastChild && (lastChild === null || lastChild === void 0 || (_lastChild$tagName = lastChild.tagName) === null || _lastChild$tagName === void 0 ? void 0 : _lastChild$tagName.toLowerCase()) === values[style]) {
          currentElement = lastChild;
        } else {
          let newElement = document.createElement(values[style]);
          currentElement.appendChild(newElement);
          currentElement = newElement;
        }
      });
    }
    const blockquotes = currentElement.querySelectorAll('blockquote');
    const listaOl = currentElement.querySelectorAll('ol');
    const listaUl = currentElement.querySelectorAll('ul');
    const codes = currentElement.querySelectorAll('code');
    if (hasContentRight) {
      isCursorInAStyleElement();
      var insertionRange = document.createRange();
      insertionRange.setStart(range.startContainer, range.startOffset);
      insertionRange.collapse(true);
      var textNode = document.createTextNode(text);
      var childElement = range.startContainer;
      if (childElement.nodeType === Node.TEXT_NODE) {
        childElement = childElement.parentNode;
      }

      // if (userPressedEnter && childElement.textContent?.length > 0 && childElement?.nodeName === 'CODE') {
      //     textNode.appendData('\n');
      // }

      const {
        exist
      } = isCursorInsideEmOrStrong();
      activeStyles === null || activeStyles === void 0 || activeStyles.forEach(style => {
        var _editorRef$current$in;
        var elements = childElement.querySelectorAll(values[style]);
        var element;
        if (elements.length > 0) {
          element = elements[elements.length - 1];
        } else {
          if (!exist) element = document.createElement(values[style]);
        }
        if (!exist && ((_editorRef$current$in = editorRef.current.innerText) === null || _editorRef$current$in === void 0 ? void 0 : _editorRef$current$in.length) > 0) {
          element = document.createElement(values[style]);
        }
        if (element) {
          element.appendChild(textNode);
          childElement.appendChild(element);
          textNode = element;
        }
      });
      insertionRange.insertNode(textNode);
      range.setStartAfter(textNode);
      range.collapse(true);
      selection.removeAllRanges();
      selection.addRange(range);
    } else {
      var _currentNode, _currentNode2;
      if (listaOl !== null && listaOl !== void 0 && listaOl.length || listaUl !== null && listaUl !== void 0 && listaUl.length) {
        var _currentList, _currentList2, _activeStyles$filter2, _listItem, _listItem2;
        const currentList = activeStyles !== null && activeStyles !== void 0 && activeStyles.includes('listOl') ? listaOl : activeStyles !== null && activeStyles !== void 0 && activeStyles.includes('listUl') ? listaUl : [];
        let listItem = (_currentList = currentList[(currentList === null || currentList === void 0 ? void 0 : currentList.length) - 1]) === null || _currentList === void 0 ? void 0 : _currentList.lastElementChild;
        let childElement;
        let brList = (_currentList2 = currentList[(currentList === null || currentList === void 0 ? void 0 : currentList.length) - 1]) === null || _currentList2 === void 0 ? void 0 : _currentList2.querySelectorAll('br');
        brList.forEach(br => {
          br.remove();
        });
        if (!currentNode) currentNode = document.createTextNode('');
        activeStyles === null || activeStyles === void 0 || (_activeStyles$filter2 = activeStyles.filter(style => !['listOl', 'listUl'].includes(style))) === null || _activeStyles$filter2 === void 0 || _activeStyles$filter2.forEach(style => {
          var _editorRef$current$in2;
          var elements = listItem.querySelectorAll(values[style]);
          var childElement;
          if (elements.length > 0) {
            childElement = elements[elements.length - 1];
          } else {
            childElement = document.createElement(values[style]);
            listItem.appendChild(childElement);
          }
          if (!isCursorInsideEmOrStrong() && ((_editorRef$current$in2 = editorRef.current.innerText) === null || _editorRef$current$in2 === void 0 ? void 0 : _editorRef$current$in2.length) > 0) {
            childElement = document.createElement(values[style]);
            listItem.appendChild(childElement);
          }
          listItem = childElement;
        });
        if ((_listItem = listItem) !== null && _listItem !== void 0 && _listItem.lastChild && ((_listItem2 = listItem) === null || _listItem2 === void 0 || (_listItem2 = _listItem2.lastChild) === null || _listItem2 === void 0 ? void 0 : _listItem2.nodeType) === Node.TEXT_NODE) {
          currentNode = listItem.lastChild;
        } else {
          currentNode = document.createTextNode('');
        }
        currentNode.appendData(text);
        listItem.appendChild(currentNode);
        if (childElement) listItem.appendChild(childElement);
      } else if (codes !== null && codes !== void 0 && codes.length) {
        var _code$lastChild;
        const code = codes[(codes === null || codes === void 0 ? void 0 : codes.length) - 1];
        const spans = code.querySelectorAll('span');
        if (spans !== null && spans !== void 0 && spans.length) {
          spans.forEach(spanElement => {
            spanElement.remove();
          });
        }
        currentNode = (code === null || code === void 0 || (_code$lastChild = code.lastChild) === null || _code$lastChild === void 0 ? void 0 : _code$lastChild.nodeType) === Node.TEXT_NODE ? code.lastChild : document.createTextNode('');
        if (userPressedEnter) {
          currentNode.appendData('\n');
        }
        currentNode.appendData(text);
        code.appendChild(currentNode);
      } else if (blockquotes !== null && blockquotes !== void 0 && blockquotes.length) {
        var _activeStyles$filter3, _childElement$lastChi;
        let paragraph = blockquotes[(blockquotes === null || blockquotes === void 0 ? void 0 : blockquotes.length) - 1].lastElementChild;
        let childElement;
        let brList = paragraph.querySelectorAll('br');
        brList.forEach(br => {
          br.remove();
        });
        if (!currentNode) currentNode = document.createTextNode('');
        activeStyles === null || activeStyles === void 0 || (_activeStyles$filter3 = activeStyles.filter(style => style !== 'quote')) === null || _activeStyles$filter3 === void 0 || _activeStyles$filter3.forEach(style => {
          var _editorRef$current$in3;
          var elements = paragraph.querySelectorAll(values[style]);
          var childElement;
          if (elements.length > 0) {
            childElement = elements[elements.length - 1];
          } else {
            childElement = document.createElement(values[style]);
            paragraph.appendChild(childElement);
          }
          if (!isCursorInsideEmOrStrong() && ((_editorRef$current$in3 = editorRef.current.innerText) === null || _editorRef$current$in3 === void 0 ? void 0 : _editorRef$current$in3.length) > 0) {
            childElement = document.createElement(values[style]);
          }
          paragraph = childElement;
        });
        if (childElement !== null && childElement !== void 0 && childElement.lastChild && (childElement === null || childElement === void 0 || (_childElement$lastChi = childElement.lastChild) === null || _childElement$lastChi === void 0 ? void 0 : _childElement$lastChi.nodeType) === Node.TEXT_NODE) {
          currentNode = childElement.lastChild;
        } else {
          currentNode = document.createTextNode('');
        }
        currentNode.appendData(text);
        paragraph.appendChild(currentNode);
        if (childElement) paragraph.appendChild(childElement);
      } else {
        if (currentElement.lastChild && currentElement.lastChild.nodeType === Node.TEXT_NODE) {
          currentNode = currentElement.lastChild;
        } else {
          currentNode = document.createTextNode('');
          currentElement.appendChild(currentNode);
        }
        currentNode.appendData(text);
      }
      var newRange = document.createRange();
      newRange.setStart(currentNode, (_currentNode = currentNode) === null || _currentNode === void 0 ? void 0 : _currentNode.length);
      newRange.setEnd(currentNode, (_currentNode2 = currentNode) === null || _currentNode2 === void 0 ? void 0 : _currentNode2.length);
      selection.removeAllRanges();
      selection.addRange(newRange);
    }
    getCaretCharacterOffsetWithin();
    removeEmptyTags(editorRef.current);
  }
  function getCaretCharacterOffsetWithin() {
    const sel = window.getSelection();
    if (sel.rangeCount > 0) {
      var _textContent$substrin;
      const range = sel.getRangeAt(0);
      const preCaretRange = range.cloneRange();
      preCaretRange.selectNodeContents(editorRef.current);
      preCaretRange.setEnd(range.endContainer, range.endOffset);
      const textOffset = preCaretRange.toString().length;
      const textContent = editorRef.current.textContent;
      setHasContentRight((textContent === null || textContent === void 0 || (_textContent$substrin = textContent.substring(textOffset, textContent === null || textContent === void 0 ? void 0 : textContent.length)) === null || _textContent$substrin === void 0 ? void 0 : _textContent$substrin.length) > 0);
    }
  }
  function isCursorInsideEmOrStrong() {
    var selection = window.getSelection();
    if (selection.rangeCount > 0) {
      var range = selection.getRangeAt(0);
      var ancestorNode = range.commonAncestorContainer;
      const tags = ['EM', 'STRONG', 'U', 'S', 'BLOCKQUOTE', 'LI', 'CODE', 'P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6'];
      while (ancestorNode && ancestorNode !== editorRef.current) {
        if (tags.includes(ancestorNode.tagName)) {
          return {
            exist: true,
            tagName: ancestorNode.tagName,
            element: ancestorNode
          };
        }
        ancestorNode = ancestorNode.parentNode;
      }
    }
    return {
      exist: false,
      tagName: null,
      element: null
    };
  }
  function removeEmptyTags(element) {
    element.querySelectorAll('*').forEach(node => {
      if (node.nodeName === 'P' && node.childNodes.length === 1 && node.firstChild.nodeName === 'BR' || node.nodeName === 'IMG') return;
      if (node.nodeName === 'BR') return;
      if (!node.innerText.trim() && !node.childNodes.length) {
        node.remove();
      }
    });
  }
  const insertParagraphWithBreak = event => {
    editorRef.current.focus();
    const selection = window.getSelection();
    const range = selection.getRangeAt(0);
    const currentNode = range.startContainer;
    if (currentNode.nodeName === 'LI') return;
    event === null || event === void 0 || event.preventDefault();
    const paragraph = document.createElement("p");
    paragraph.appendChild(document.createElement("br"));
    if (!range.collapsed) {
      range.deleteContents();
    }
    range.insertNode(paragraph);
    const newRange = document.createRange();
    newRange.setStartAfter(paragraph);
    newRange.collapse(true);
    selection.removeAllRanges();
    selection.addRange(newRange);
  };
  function addLink() {
    editorRef.current.focus();
    const selection = window.getSelection();
    const range = selection.getRangeAt(0);
    setSavedRange(range);
    createLinkRef.current.style.display = 'block';
    createLinkRef.current.firstChild.focus();
  }
  const createLink = () => {
    editorRef.current.focus();
    const range = savedRange || window.getSelection().getRangeAt(0);
    const tags = ['EM', 'STRONG', 'U', 'S', 'BLOCKQUOTE', 'OL', 'UL', 'CODE', 'P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6'];
    let currentNode = range.startContainer;
    let isCursorInAllowedElement = false;
    while (currentNode && currentNode !== editorRef.current) {
      if (tags.includes(currentNode.tagName)) {
        isCursorInAllowedElement = true;
        break;
      }
      currentNode = currentNode.parentNode;
    }

    // Criar o link
    const link = document.createElement("a");
    link.href = linkInfo.url;
    link.textContent = linkInfo.title;

    // Se o cursor não estiver dentro de um dos elementos permitidos, criar um parágrafo
    if (!isCursorInAllowedElement) {
      const paragraph = document.createElement('p');
      paragraph.appendChild(link);
      paragraph.appendChild(range.extractContents());
      range.insertNode(paragraph);
    } else {
      // Inserir o link na posição atual do cursor
      range.insertNode(link);
    }

    // Limpar os dados do link e esconder o formulário de criação de link
    setLinkInfo({
      url: '',
      title: ''
    });
    createLinkRef.current.style.display = 'none';
  };
  const insertImage = imageUrl => {
    const imageContainer = document.createElement('div');
    const image = document.createElement('img');
    imageContainer.setAttribute('class', 'image-container');
    image.src = imageUrl;
    imageContainer.appendChild(image);
    editorRef.current.appendChild(imageContainer);
  };
  function resetStyles() {
    setTextColor('#162949');
    let newStyles = {
      bold: false,
      italic: false,
      underline: false,
      strikethrough: false,
      h1: false,
      h2: false,
      h3: false,
      h4: false,
      h5: false,
      h6: false
    };
    const selection = window.getSelection();
    const range = selection.getRangeAt(0);
    const currentNode = range.startContainer;
    if (currentNode.nodeName === 'LI') {
      newStyles = {
        ...newStyles,
        listOl: false,
        listUl: false
      };
    }
    setStyles(prevStyles => {
      return {
        ...prevStyles,
        ...newStyles
      };
    });
  }
  const defineCorDoTexto = value => {
    setTextColor(value);
  };
  const handleImageUpload = file => {
    if (!file) return;
    insertImage(URL.createObjectURL(file));
  };
  const handleClickImageButton = () => {
    const fileInput = document.getElementById('fileInput');
    fileInput.click();
  };
  const handleClick = event => {
    var _editorRef$current$in4;
    if (!((_editorRef$current$in4 = editorRef.current.innerText) !== null && _editorRef$current$in4 !== void 0 && _editorRef$current$in4.length)) return;
    const rect = event.target.getBoundingClientRect();
    const y = event.clientY - rect.top;
    if (y <= rect.height - 20) return;
    let newStyles = {
      bold: false,
      italic: false,
      underline: false,
      strikethrough: false,
      h1: false,
      h2: false,
      h3: false,
      h4: false,
      h5: false,
      h6: false,
      listOl: false,
      listUl: false
    };
    setStyles(newStyles);
    limpaEstilos();
    const p = document.createElement("p");
    p.innerHTML = '<br>';
    const sel = window.getSelection();
    const range = document.createRange();
    range.selectNodeContents(editorRef.current);
    range.collapse(false);
    range.insertNode(p);
    range.setStart(p, 0);
    range.collapse(true);
    sel.removeAllRanges();
    sel.addRange(range);
  };
  const limpaEstilos = () => {
    var _document$getElementB, _document$getElementB2, _document$getElementB3, _document$getElementB4, _document$getElementB5, _document$getElementB6;
    (_document$getElementB = document.getElementById('boldButton')) === null || _document$getElementB === void 0 || (_document$getElementB = _document$getElementB.classList) === null || _document$getElementB === void 0 || _document$getElementB.remove('active');
    (_document$getElementB2 = document.getElementById('italicButton')) === null || _document$getElementB2 === void 0 || (_document$getElementB2 = _document$getElementB2.classList) === null || _document$getElementB2 === void 0 || _document$getElementB2.remove('active');
    (_document$getElementB3 = document.getElementById('underlineButton')) === null || _document$getElementB3 === void 0 || (_document$getElementB3 = _document$getElementB3.classList) === null || _document$getElementB3 === void 0 || _document$getElementB3.remove('active');
    (_document$getElementB4 = document.getElementById('strikethroughButton')) === null || _document$getElementB4 === void 0 || (_document$getElementB4 = _document$getElementB4.classList) === null || _document$getElementB4 === void 0 || _document$getElementB4.remove('active');
    (_document$getElementB5 = document.getElementById('listOlButton')) === null || _document$getElementB5 === void 0 || (_document$getElementB5 = _document$getElementB5.classList) === null || _document$getElementB5 === void 0 || _document$getElementB5.remove('active');
    (_document$getElementB6 = document.getElementById('listUlButton')) === null || _document$getElementB6 === void 0 || (_document$getElementB6 = _document$getElementB6.classList) === null || _document$getElementB6 === void 0 || _document$getElementB6.remove('active');
  };
  const isCursorInAStyleElement = () => {
    const selection = window.getSelection();
    const buttons = {
      strong: 'bold',
      em: 'italic',
      u: 'underline',
      s: 'strikethrough',
      ol: 'listOl',
      ul: 'listUl',
      blockquote: 'quote',
      code: 'code'
    };
    if (selection.rangeCount > 0) {
      var range = savedRange || selection.getRangeAt(0);
      var ancestorNode = range.commonAncestorContainer;
      const tags = ['EM', 'STRONG', 'U', 'S', 'BLOCKQUOTE', 'UL', 'OL', 'CODE'];
      let foundTags = [];
      while (ancestorNode && ancestorNode !== editorRef.current) {
        if (tags.includes(ancestorNode.tagName)) {
          foundTags.push(ancestorNode.tagName.toLowerCase());
        }
        ancestorNode = ancestorNode.parentNode;
      }
      limpaEstilos();
      foundTags.forEach(tag => {
        const button = document.getElementById("".concat(buttons[tag], "Button"));
        if (button) {
          button.classList.add('active');
          setStyles(prevStyles => {
            return {
              ...prevStyles,
              [buttons[tag]]: true
            };
          });
        }
      });
    }
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "b-editor-container",
    id: "b-editor-container",
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("main", {
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        id: "toolbar",
        className: "b-editor-container__toolbar",
        children: [screenSize > 767 && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "b-editor-container__toolbar__group",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: fontSelecionada,
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: opcoesDeFontes === null || opcoesDeFontes === void 0 ? void 0 : opcoesDeFontes.map((font, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => selecionaFonteDoTexto(font),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    className: "b-editor-container__toolbar__group--font b-editor-container__toolbar__group--font--".concat(font.value),
                    children: font.label
                  })
                }, index))
              })
            })
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "b-editor-container__toolbar__group",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "boldButton",
              onClick: () => toggleButton('bold', 'boldButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaBold, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "italicButton",
              onClick: () => toggleButton('italic', 'italicButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaItalic, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "underlineButton",
              onClick: () => toggleButton('underline', 'underlineButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaUnderline, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "strikethroughButton",
              onClick: () => toggleButton('strikethrough', 'strikethroughButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaStrikethrough, {})
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "b-editor-container__toolbar__group",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "listOlButton",
              onClick: () => toggleButton('listOl', 'listOlButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa2.FaListOl, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "listUlButton",
              onClick: () => toggleButton('listUl', 'listUlButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaListUl, {})
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "b-editor-container__toolbar__group",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "linkButton",
              onClick: addLink,
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaLink, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "codeButton",
              onClick: () => toggleButton('code', 'codeButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaCode, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "quoteButton",
              onClick: () => toggleButton('quote', 'quoteButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaQuoteRight, {})
            })]
          })]
        }), screenSize > 505 && screenSize < 767 && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "b-editor-container__toolbar__group",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: "Aa",
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: opcoesDeFontes === null || opcoesDeFontes === void 0 ? void 0 : opcoesDeFontes.map((font, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => selecionaFonteDoTexto(font),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    className: "b-editor-container__toolbar__group--font b-editor-container__toolbar__group--font--".concat(font.value),
                    children: font.label
                  })
                }, index))
              })
            })
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "b-editor-container__toolbar__group",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "boldButton",
              onClick: () => toggleButton('bold', 'boldButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaBold, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "italicButton",
              onClick: () => toggleButton('italic', 'italicButton'),
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaItalic, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaEllipsisH, {})
              }),
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              hideDropdownIcon: true,
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: (_ref2 = [{
                  label: 'Sublinhado',
                  action: 'underline'
                }, {
                  label: 'Tachado',
                  action: 'strikethrough'
                }]) === null || _ref2 === void 0 ? void 0 : _ref2.map((style, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => toggleButton(style.action, "".concat(style.action, "Button")),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    children: style.label
                  })
                }, index))
              })
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "b-editor-container__toolbar__group",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa2.FaListOl, {})
              }),
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: (_ref3 = [{
                  label: 'Lista de marcadores',
                  action: 'listOl'
                }, {
                  label: 'Lista numerada',
                  action: 'listUl'
                }]) === null || _ref3 === void 0 ? void 0 : _ref3.map((style, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => toggleButton(style.action, "".concat(style.action, "Button")),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    children: style.label
                  })
                }, index))
              })
            })
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "b-editor-container__toolbar__group",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              id: "linkButton",
              onClick: addLink,
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaLink, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaPlus, {})
              }),
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: (_ref4 = [{
                  label: 'Trecho de código',
                  action: 'code'
                }, {
                  label: 'Citação',
                  action: 'quote'
                }]) === null || _ref4 === void 0 ? void 0 : _ref4.map((style, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => toggleButton(style.action, "".concat(style.action, "Button")),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    children: style.label
                  })
                }, index))
              })
            })]
          })]
        }), screenSize < 505 && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "b-editor-container__toolbar__group",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: "Aa",
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: opcoesDeFontes === null || opcoesDeFontes === void 0 ? void 0 : opcoesDeFontes.map((font, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => selecionaFonteDoTexto(font),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    className: "b-editor-container__toolbar__group--font b-editor-container__toolbar__group--font--".concat(font.value),
                    children: font.label
                  })
                }, index))
              })
            })
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "b-editor-container__toolbar__group",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaBold, {})
              }),
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: (_ref5 = [{
                  label: 'Negrito',
                  action: 'bold'
                }, {
                  label: 'Itálico',
                  action: 'italic'
                }, {
                  label: 'Sublinhado',
                  action: 'underline'
                }, {
                  label: 'Tachado',
                  action: 'strikethrough'
                }]) === null || _ref5 === void 0 ? void 0 : _ref5.map((style, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => toggleButton(style.action, "".concat(style.action, "Button")),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    children: style.label
                  })
                }, index))
              })
            })
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "b-editor-container__toolbar__group",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa2.FaListOl, {})
              }),
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: (_ref6 = [{
                  label: 'Lista de marcadores',
                  action: 'listOl'
                }, {
                  label: 'Lista numerada',
                  action: 'listUl'
                }]) === null || _ref6 === void 0 ? void 0 : _ref6.map((style, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => toggleButton(style.action, "".concat(style.action, "Button")),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    children: style.label
                  })
                }, index))
              })
            })
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "b-editor-container__toolbar__group",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.Dropdown, {
              trigger: /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
                children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_fa.FaPlus, {})
              }),
              triggerClassName: 'b-editor-container__toolbar__group--selector',
              placement: "bottom",
              children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownContent, {
                children: (_ref7 = [{
                  label: 'Link',
                  action: 'code',
                  fn: addLink
                }, {
                  label: 'Adicionar imagem',
                  action: 'code',
                  fn: handleClickImageButton
                }, {
                  label: 'Trecho de código',
                  action: 'code'
                }, {
                  label: 'Citação',
                  action: 'quote'
                }]) === null || _ref7 === void 0 ? void 0 : _ref7.map((style, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)(_Dropdown.DropdownItem, {
                  onClick: () => style !== null && style !== void 0 && style.fn ? style.fn : toggleButton(style.action, "".concat(style.action, "Button")),
                  children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
                    children: style.label
                  })
                }, index))
              })
            })
          })]
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: "b-editor-container__content",
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "b-editor-container__content--editor",
          ref: editorRef,
          contentEditable: true,
          suppressContentEditableWarning: true,
          "data-placeholder": "Digite algo aqui...",
          onKeyPress: handleKeyPress,
          onInput: () => {
            var _editorRef$current;
            return editorRef === null || editorRef === void 0 || (_editorRef$current = editorRef.current) === null || _editorRef$current === void 0 ? void 0 : _editorRef$current.focus();
          },
          onClick: () => {
            getCaretCharacterOffsetWithin();
            isCursorInAStyleElement();
          }
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          onClick: handleClick,
          className: "b-editor-container__content--click"
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: "create-link",
        ref: createLinkRef,
        children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "create-link__content",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "create-link__content__field",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("label", {
              htmlFor: "",
              className: "create-link__content__field--label",
              children: "Texto para exibi\xE7\xE3o"
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
              type: "text",
              name: "",
              id: "",
              className: "create-link__content__field--input",
              value: linkInfo === null || linkInfo === void 0 ? void 0 : linkInfo.title,
              onChange: event => setLinkInfo({
                ...linkInfo,
                title: event.target.value
              })
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: "create-link__content__field",
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("label", {
              htmlFor: "",
              className: "create-link__content__field--label",
              children: "Link"
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
              type: "text",
              name: "",
              id: "",
              className: "create-link__content__field--input",
              value: linkInfo === null || linkInfo === void 0 ? void 0 : linkInfo.url,
              onChange: event => setLinkInfo({
                ...linkInfo,
                url: event.target.value
              })
            })]
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "create-link__content__field",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
              className: "create-link__content__field--button",
              onClick: createLink,
              children: "Save"
            })
          })]
        })
      })]
    })
  });
}
