// powerful-react-table/dist/components/Tooltip/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Tooltip;
var _react = _interopRequireWildcard(require("react"));
var _TooltipModule = _interopRequireDefault(require("./Tooltip.module.scss"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function Tooltip(_ref) {
  let {
    content,
    position,
    customStyle,
    children,
    hidden
  } = _ref;
  const tooltipRef = (0, _react.useRef)();
  const containerRef = (0, _react.useRef)();
  const handleDisplayTooltip = () => {
    if (!tooltipRef?.current) return;
    tooltipRef.current.style.display = 'block';
    const tooltipRect = tooltipRef?.current?.getBoundingClientRect();
    const containerRect = containerRef.current.getBoundingClientRect();
    const spaceRight = window.innerWidth - containerRect.right;
    if (['top', 'bottom'].includes(position)) {
      if (spaceRight >= tooltipRect.width && containerRect.left - containerRect.width < tooltipRect.width) tooltipRef.current.style.left = '0';
      if (spaceRight <= tooltipRect.width && containerRect.left > tooltipRect.width) tooltipRef.current.style.right = '0';
    }
  };
  const handleHideTooltip = () => {
    if (tooltipRef?.current) tooltipRef.current.style.display = 'none';
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    ref: containerRef,
    className: _TooltipModule.default['tooltip-container'],
    onMouseEnter: handleDisplayTooltip,
    onMouseLeave: handleHideTooltip,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: _TooltipModule.default['tooltip-container__content'],
      children: [children, !hidden && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        ref: tooltipRef,
        className: `${_TooltipModule.default['tooltip-container__tooltip']} ${_TooltipModule.default[`tooltip-container__tooltip--${position || 'bottom'}`]}`,
        style: customStyle,
        children: content
      })]
    })
  });
}
