// react-view-files/dist/components/ContentViewer/index.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ContentViewer;
var _react = _interopRequireWildcard(require("react"));
var _md = require("react-icons/md");
var _ContentViewerModule = _interopRequireDefault(require("./ContentViewer.module.scss"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function ContentViewer(_ref) {
  var children = _ref.children,
    file = _ref.file,
    type = _ref.type,
    visible = _ref.visible,
    onClose = _ref.onClose;
  var _useState = (0, _react.useState)(false),
    _useState2 = _slicedToArray(_useState, 2),
    contentVisible = _useState2[0],
    setContentVisible = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    fileExtension = _useState4[0],
    setFileExtension = _useState4[1];
  var _useState5 = (0, _react.useState)(true),
    _useState6 = _slicedToArray(_useState5, 2),
    headerVisible = _useState6[0],
    setHeaderVisible = _useState6[1]; // Estado para controlar a visibilidade do header
  var containerRef = (0, _react.useRef)(null);
  var interactionTimer = (0, _react.useRef)(null); // Referência para o timer de interação

  var icons = {
    pdf: require("../../img/pdf.png"),
    excel: require("../../img/excel.png"),
    image: require("../../img/picture.png"),
    word: require("../../img/word.png"),
    audio: require("../../img/audio.png"),
    video: require("../../img/play.png"),
    txt: require("../../img/news.png")
  };
  var loadFiles = ["pdf", "image", "audio", "video", "word", "txt"];
  (0, _react.useEffect)(function () {
    setContentVisible(visible);
    if (visible) document.body.style.overflowY = "hidden";
  }, [visible]);
  (0, _react.useEffect)(function () {
    return setFileExtension(type);
  }, [type]);

  // Função para ocultar o cabeçalho
  var hideHeader = function hideHeader() {
    setHeaderVisible(false);
  };

  // Função para resetar o timer de interação
  var resetInteractionTimer = function resetInteractionTimer() {
    setHeaderVisible(true);
    if (interactionTimer.current) clearTimeout(interactionTimer.current);
    interactionTimer.current = setTimeout(hideHeader, 2500);
  };

  // Adiciona listeners de interação ao montar o componente
  (0, _react.useEffect)(function () {
    window.addEventListener("mousemove", resetInteractionTimer);
    window.addEventListener("keydown", resetInteractionTimer);

    // Limpa os listeners ao desmontar o componente
    return function () {
      window.removeEventListener("mousemove", resetInteractionTimer);
      window.removeEventListener("keydown", resetInteractionTimer);
      if (interactionTimer.current) clearTimeout(interactionTimer.current);
    };
  }, []);
  (0, _react.useEffect)(function () {
    window.addEventListener("keydown", function (event) {
      if (event.code === "Escape") {
        setContentVisible(false);
        onClose === null || onClose === void 0 || onClose();
        setFileExtension(null);
        resetOverflowY();
      }
    });
  });
  var resetOverflowY = function resetOverflowY() {
    return document.body.style.overflowY = "auto";
  };
  var downloadFile = function downloadFile() {
    var url = file !== null && file !== void 0 && file.url ? file.url : window.URL.createObjectURL(file);
    var link = document.createElement("a");
    link.download = file.name;
    link.href = file.url || url;
    link.click();
    window.URL.revokeObjectURL(url);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "".concat(_ContentViewerModule["default"]["container-viewer"], " ").concat(contentVisible ? _ContentViewerModule["default"]["container-viewer--show"] : null),
    ref: containerRef,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "".concat(_ContentViewerModule["default"]["container-viewer__header"], " ").concat(!(loadFiles !== null && loadFiles !== void 0 && loadFiles.includes(fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type)) ? _ContentViewerModule["default"]["container-viewer__header--dark"] : "", " ").concat(headerVisible ? '' : _ContentViewerModule["default"]["container-viewer__header--hidden"]) // Adiciona classe para ocultar
      ,
      children: [fileExtension !== null && fileExtension !== void 0 && fileExtension.name ? /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: "d-flex align-items-center",
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: _ContentViewerModule["default"]["file__icon"],
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
            src: icons[fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type],
            alt: file === null || file === void 0 ? void 0 : file.name
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "pl-1",
          children: [file === null || file === void 0 ? void 0 : file.name, ".", fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.name]
        })]
      }) : /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {}), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: _ContentViewerModule["default"]["viewer-actions"],
        children: [(loadFiles === null || loadFiles === void 0 ? void 0 : loadFiles.includes(fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type)) && /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: _ContentViewerModule["default"]["viewer-actions__link"],
          onClick: function onClick() {
            return downloadFile();
          },
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineFileDownload, {})
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: _ContentViewerModule["default"]["viewer-actions__link"],
          onClick: function onClick() {
            resetOverflowY();
            setContentVisible(false);
            onClose === null || onClose === void 0 || onClose();
            setFileExtension(null);
          },
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineClose, {})
        })]
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: _ContentViewerModule["default"]["container-viewer__content"],
      id: "viewer-content",
      children: children
    }), !(loadFiles !== null && loadFiles !== void 0 && loadFiles.includes(fileExtension === null || fileExtension === void 0 ? void 0 : fileExtension.type)) && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: _ContentViewerModule["default"]["container-viewer__no-file"],
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        style: {
          marginBottom: "20px"
        },
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)("img", {
          src: require("../../img/corrupted-file.png"),
          alt: "Corrupted File"
        })
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        children: "Algo deu errado."
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        children: "N\xE3o foi poss\xEDvel gerar uma visualiza\xE7\xE3o para este arquivo."
      })]
    })]
  });
}
