// react-view-files/dist/components/Modal/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalHeader = exports.ModalFooter = exports.ModalBody = exports.Modal = void 0;
var _react = _interopRequireDefault(require("react"));
var _ModalModule = _interopRequireDefault(require("./Modal.module.scss"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
var ModalOverlay = function ModalOverlay() {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "".concat(_ModalModule["default"]['modal-overlay'])
  });
};
var ModalDialogContent = function ModalDialogContent(_ref) {
  var children = _ref.children,
    options = _ref.options;
  var size = options.size;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "".concat(_ModalModule["default"]['dialog']),
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "".concat(_ModalModule["default"]['content'], " ").concat(size ? _ModalModule["default"][size] : _ModalModule["default"]['small']),
      children: children
    })
  });
};
var Modal = exports.Modal = function Modal(_ref2, props) {
  var show = _ref2.show,
    content = _ref2.content,
    alignment = _ref2.alignment,
    children = _ref2.children;
  var theme = 'light';
  if (show) {
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "".concat(_ModalModule["default"]['modal-container'], " ").concat(alignment ? _ModalModule["default"]["flex-".concat(alignment)] : _ModalModule["default"]['flex-center'], " ").concat(content ? _ModalModule["default"]["content-".concat(content)] : '', " ").concat(_ModalModule["default"][theme]),
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(ModalOverlay, {}), /*#__PURE__*/(0, _jsxRuntime.jsx)(ModalDialogContent, {
        options: props,
        children: children
      })]
    });
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_jsxRuntime.Fragment, {});
};
var ModalHeader = exports.ModalHeader = function ModalHeader(_ref3) {
  var children = _ref3.children,
    closeButton = _ref3.closeButton,
    onClose = _ref3.onClose;
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "".concat(_ModalModule["default"]['header__modal']),
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("h4", {
      className: "".concat(_ModalModule["default"]['modal__title']),
      children: children
    }), closeButton && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      className: "material-icons ".concat(_ModalModule["default"]['close__modal']),
      onClick: onClose,
      children: "close"
    })]
  });
};
var ModalBody = exports.ModalBody = function ModalBody(_ref4) {
  var children = _ref4.children;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "".concat(_ModalModule["default"]['body']),
    children: children
  });
};
var ModalFooter = exports.ModalFooter = function ModalFooter(_ref5) {
  var children = _ref5.children;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "".concat(_ModalModule["default"]['footer-modal']),
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "".concat(_ModalModule["default"]['footer-modal__buttons']),
      children: children
    })
  });
};
