// powerful-react-table/dist/components/Calendar/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Calendar;
var _react = _interopRequireWildcard(require("react"));
var _moment = _interopRequireDefault(require("moment"));
var _uuid = require("uuid");
var _md = require("react-icons/md");
require("./Calendar.scss");
var _localeConfig = require("../../localeConfig");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function Calendar(_ref) {
  let {
    onChange
  } = _ref;
  const [state, setState] = (0, _react.useState)({
    month: (0, _moment.default)(),
    selected: (0, _moment.default)().startOf('day'),
    dates: {
      start: null,
      end: null
    }
  });
  const userLocale = (0, _localeConfig.getUserLocale)()?.locale;
  const previous = () => {
    const {
      month
    } = state;
    setState({
      ...state,
      month: month.subtract(1, 'month')
    });
  };
  const next = () => {
    const {
      month
    } = state;
    setState({
      ...state,
      month: month.add(1, 'month')
    });
  };
  const select = (day, event) => {
    setState({
      ...state,
      selected: day.date,
      month: day.date.clone()
    });
    onChange?.(day.date.format(userLocale === 'en-US' ? 'MM/DD/YYYY' : 'DD/MM/YYYY'));
  };
  const getWeeks = () => {
    let weeks = [];
    let concluded = false;
    let date = state.month.clone().startOf('month').add('w' - 1).day('Sunday');
    let count = 0;
    let monthIndex = date.month();
    const {
      selected,
      month
    } = state;
    while (!concluded) {
      weeks.push(/*#__PURE__*/(0, _jsxRuntime.jsx)(Week, {
        date: date.clone(),
        month: month,
        select: (day, event) => select(day, event),
        selected: selected
      }, (0, _uuid.v4)()));
      date.add(1, 'w');
      concluded = count++ > 2 && monthIndex !== date.month();
      monthIndex = date.month();
    }
    return weeks;
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: 'prt-calendar__container',
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: 'prt-calendar__container__header',
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: 'header__month',
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: 'header__month__action',
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdNavigateBefore, {
            onClick: () => previous()
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            className: `size-16 weight-500`,
            style: {
              textTransform: 'capitalize'
            },
            children: state.month.format('MMMM YYYY')
          })
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: 'header__month__action',
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdNavigateNext, {
            onClick: () => next()
          })
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(Days, {})]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: 'prt-calendar__container__content',
      children: getWeeks()
    })]
  });
}
const Days = () => {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: 'weekly__days',
    children: ['Dom', 'Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sáb'].map((day, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      className: 'day',
      children: day
    }, index))
  });
};
const Week = props => {
  let days = [];
  let {
    date
  } = props;
  const {
    month,
    selected,
    select
  } = props;
  for (var i = 0; i < 7; i++) {
    let day = {
      name: date.format('dd').substring(0, 1),
      number: date.date(),
      isCurrentMonth: date.month() === month.month(),
      isToday: date.isSame(new Date(), 'day'),
      date: date
    };
    days.push(/*#__PURE__*/(0, _jsxRuntime.jsx)(Day, {
      day: day,
      selected: selected,
      select: select
    }, (0, _uuid.v4)()));
    date = date.clone();
    date.add(1, 'day');
  }
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: 'week',
    children: days
  }, days[0]);
};
const Day = props => {
  const {
    day,
    day: {
      date,
      isCurrentMonth,
      isToday,
      number
    },
    select,
    selected
  } = props;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: `day ${isToday ? 'today' : ''} ${isCurrentMonth ? '' : 'another__month'} ${date.isSame(selected) ? 'selected' : ''}`,
    onClick: event => select(day, event),
    children: number
  }, (0, _uuid.v4)());
};
