// react-view-files/dist/components/DropdownMenuOptions/index.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = DropdownMenuOptions;
var _react = _interopRequireWildcard(require("react"));
var _useOut = _interopRequireDefault(require("../../useOut"));
var _md = require("react-icons/md");
require("./DropdownMenuOptions.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function DropdownMenuOptions(_ref) {
  var options = _ref.options,
    onClicked = _ref.onClicked,
    direction = _ref.direction;
  var _useState = (0, _react.useState)(false),
    _useState2 = _slicedToArray(_useState, 2),
    dropdownOpen = _useState2[0],
    setDropdownOpen = _useState2[1];
  var _useState3 = (0, _react.useState)({
      top: 0,
      left: 0
    }),
    _useState4 = _slicedToArray(_useState3, 2),
    position = _useState4[0],
    setPosition = _useState4[1];
  var dropdownRef = (0, _react.useRef)(null);
  var dropdownContentRef = (0, _react.useRef)(null);
  var childRef = (0, _react.useRef)(null);
  (0, _useOut["default"])(dropdownRef, function () {
    setDropdownOpen(false);
  });
  var handleToggleDropdown = function handleToggleDropdown(event) {
    if (dropdownOpen) {
      return setDropdownOpen(false);
    }
    calculatePosition(event);
    setDropdownOpen(true);
  };
  function navigateTo(url, onNavigateCallback) {
    try {
      window.location.href = url;
      if (onNavigateCallback) {
        onNavigateCallback();
      }
    } catch (error) {
      console.error("Erro durante a navegação:", error);
    }
  }
  var handleNavigateItem = function handleNavigateItem(url) {
    var redirect = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
    setDropdownOpen(false);
    if (redirect) navigateTo(url);
  };
  var calculatePosition = function calculatePosition(event) {
    if (dropdownRef.current && event) {
      var _event$target;
      var parentRect = event === null || event === void 0 || (_event$target = event.target) === null || _event$target === void 0 ? void 0 : _event$target.getBoundingClientRect();
      var dropdownRect = dropdownContentRef.current.getBoundingClientRect();
      var spaceRight = window.innerWidth - parentRect.right;
      var spaceAbove = parentRect.top;
      var spaceBelow = window.innerHeight - parentRect.bottom;
      var newPosition = {
        left: spaceRight >= dropdownRect.width ? parentRect.left : parentRect.right - ((dropdownRect === null || dropdownRect === void 0 ? void 0 : dropdownRect.width) || 160)
      };
      setTimeout(function () {
        var _childRef$current$get;
        var childHeight = ((_childRef$current$get = childRef.current.getBoundingClientRect()) === null || _childRef$current$get === void 0 ? void 0 : _childRef$current$get.height) + 18;
        if (spaceAbove < spaceBelow) {
          newPosition.top = parentRect.bottom + 10;
        } else {
          newPosition.top = parentRect.top - childHeight;
        }
        setPosition(newPosition);
      }, 1);
    }
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "dmc-dropdown-menu-container",
    ref: dropdownRef,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "dropdown__header",
      onClick: function onClick(event) {
        return handleToggleDropdown(event);
      },
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineMoreHoriz, {
        className: "dmc-dropdown-icon"
      })
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "dropdown__content ".concat(dropdownOpen ? "visible" : "", " ").concat(direction === "top" ? "reverse" : ""),
      ref: dropdownContentRef,
      style: {
        top: position.top,
        left: position.left
      },
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {
        ref: childRef,
        children: options === null || options === void 0 ? void 0 : options.map(function (option, index) {
          return /*#__PURE__*/(0, _jsxRuntime.jsxs)("li", {
            className: "dropdow__item",
            onClick: function onClick() {
              handleNavigateItem(option === null || option === void 0 ? void 0 : option.url, option === null || option === void 0 ? void 0 : option.redirect);
              onClicked === null || onClicked === void 0 || onClicked(option);
            },
            children: [option.icon && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
              style: {
                fontSize: "20px",
                color: "#162949",
                display: "flex"
              },
              children: option.icon
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
              children: option.name
            })]
          }, index);
        })
      })
    })]
  });
}
