// powerful-react-table/dist/components/Table/DropdownFilter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DropdownFilter;
var _react = _interopRequireWildcard(require("react"));
var _useTable = require("../../useTable");
var _useOut = _interopRequireDefault(require("../../../useOut"));
var _Calendar = _interopRequireDefault(require("../../Calendar"));
var _Input = _interopRequireDefault(require("../../Input"));
var _md = require("react-icons/md");
var _language = _interopRequireDefault(require("../language.json"));
var _DropdownFilterModule = _interopRequireDefault(require("./DropdownFilter.module.scss"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function DropdownFilter(_ref) {
  let {
    filter,
    identifier,
    createRef,
    onClose,
    locale = 'en'
  } = _ref;
  const [type, setType] = (0, _react.useState)('');
  const [options, setOptions] = (0, _react.useState)([]);
  const [selected, setSelected] = (0, _react.useState)([]);
  const [searchField, setSearchField] = (0, _react.useState)('');
  const [between, setBetween] = (0, _react.useState)({
    min: 0,
    max: 0
  });
  const [dates, setDates] = (0, _react.useState)({
    start: null,
    end: null
  });
  const [position, setPosition] = (0, _react.useState)({
    top: 0,
    left: 0
  });
  const containerRef = (0, _react.useRef)(null);
  const dropdownRef = (0, _react.useRef)(null);
  const contentRef = (0, _react.useRef)(null);
  const {
    searchFilter,
    setSearchFilter,
    setRemoveFilter
  } = (0, _useTable.useTable)();
  const VALUES = _language.default[locale?.toUpperCase()];
  const labelOptions = {
    text: VALUES.TEXT_FILTER,
    option: VALUES.OPTION_FILTER,
    date: VALUES.DATE_FILTER,
    between: VALUES.BETWEEN_FILTER
  };
  const classes = {
    text: 'text-content',
    option: 'option-content',
    date: 'date-content',
    between: 'between-content'
  };
  const savedFilter = JSON.parse(localStorage.getItem('__magnunTablePreferences') || null)?.filters?.[identifier];
  (0, _react.useEffect)(() => setOptionValues(), [filter?.options]);
  (0, _react.useEffect)(() => {
    setType(filter?.type);
  }, [filter?.type]);
  (0, _react.useEffect)(() => {
    setSavedValues();
  }, [options]);
  (0, _react.useEffect)(() => {
    calculatePosition();
  }, []);
  (0, _react.useEffect)(() => {
    window.addEventListener('resize', calculatePosition);
    return () => {
      window.removeEventListener('resize', calculatePosition);
    };
  }, []);
  (0, _useOut.default)(containerRef, () => onClose());
  const calculatePosition = () => {
    const widths = {
      text: 350,
      option: 300,
      date: 600,
      between: 350
    };
    if (dropdownRef.current && createRef) {
      const parentRect = createRef.getBoundingClientRect();
      const dropdownRect = dropdownRef.current.getBoundingClientRect();
      const spaceRight = window.innerWidth - parentRect.right;
      const spaceAbove = parentRect.top;
      const spaceBelow = window.innerHeight - parentRect.bottom;
      let newPosition = {
        left: spaceRight >= dropdownRect.width ? parentRect.left : parentRect.right - widths[filter?.type]
      };
      setTimeout(() => {
        const containerHeight = contentRef?.current?.getBoundingClientRect()?.height + 115;
        if (spaceAbove < spaceBelow) {
          newPosition.top = parentRect.bottom + 15;
        } else {
          newPosition.top = parentRect.top - containerHeight;
        }
        setPosition(newPosition);
      }, 1);
    }
  };
  const setOptionValues = () => {
    const values = filter?.options;
    values?.forEach(option => option.checked = false);
    setOptions([...values]);
    setSelected([]);
  };
  const setSavedValues = () => {
    const saved = savedFilter ? Array.from(Object.values(savedFilter)) : [];
    saved?.forEach(item => {
      if (item?.type === 'text') filter.selected = item?.value;
      if (item?.type === 'option') selectedOptions(item);
      if (item?.type === 'between') selectedBetween(item);
    });
  };
  const selectedOptions = item => {
    if (item?.position !== filter?.position) return;
    filter?.options?.forEach(option => option.checked = item?.value?.includes(option?.value));
  };
  const selectedBetween = item => {
    if (item?.position !== filter?.position) return;
    setBetween({
      min: item?.value?.min,
      max: item?.value?.max
    });
  };
  const handleCheckboxSelect = index => {
    const hasExist = selected?.findIndex(value => value === index);
    let filtered = selected;
    if (hasExist !== -1) {
      filtered?.splice(hasExist, 1);
      setSelected(filtered);
      applySelected();
      return;
    }
    filtered.push(index);
    setSelected(filtered);
    applySelected();
  };
  const applySelected = () => {
    const items = options?.map((item, index) => ({
      ...item,
      checked: selected?.includes(index)
    }));
    setOptions(items);
  };
  const handleClear = () => {
    const storage = localStorage.getItem('__magnunTablePreferences');
    const savedFilter = JSON.parse(storage)?.filters;
    if (!savedFilter) return onClose();
    delete savedFilter[identifier][filter?.label?.toUpperCase()];
    localStorage.setItem('__magnunTablePreferences', JSON.stringify({
      ...JSON.parse(storage),
      filters: savedFilter
    }));
    setOptionValues();
    onClose();
  };
  const handleSelectAll = () => {
    const items = options?.map(item => ({
      ...item,
      checked: true
    }));
    setOptions(items);
    setSelected(options?.map((options, index) => index));
  };
  const applyFilter = () => {
    if (selected?.length) {
      changeDispatch({
        type: 'option',
        value: options?.filter((value, index) => selected?.includes(index))?.map(option => option.value)
      });
      return;
    }
    if (searchField) {
      changeDispatch({
        type: 'text',
        value: searchField
      });
      return;
    }
    if (between.min && between.max) {
      changeDispatch({
        type: 'between',
        value: between
      });
      return;
    }
    if (dates.start && dates.end) {
      changeDispatch({
        type: 'date',
        value: dates
      });
      return;
    }
  };
  const changeDispatch = value => {
    setSearchFilter([...searchFilter, {
      ...value,
      label: filter?.label,
      position: filter?.position
    }]);
    const storage = localStorage.getItem('__magnunTablePreferences');
    const filtersStorage = JSON.parse(storage)?.filters;
    const newVlues = {
      ...filtersStorage,
      [identifier]: {
        ...filtersStorage?.[identifier],
        [filter?.label?.toUpperCase()]: {
          label: filter?.label,
          position: filter?.position,
          type: value.type,
          value: value.value
        }
      }
    };
    localStorage.setItem('__magnunTablePreferences', JSON.stringify({
      ...JSON.parse(storage),
      filters: newVlues
    }));
    setRemoveFilter(false);
    onClose();
  };
  const searchForOptions = value => {
    const values = filter?.options?.filter(option => option?.name?.toLowerCase().indexOf(value.toLowerCase()) > -1);
    setOptions([...values]);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: `${_DropdownFilterModule.default['dropdown-filter-container']} ${type === 'date' ? _DropdownFilterModule.default['dropdown-filter-container--large'] : ''}`,
    ref: containerRef,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: _DropdownFilterModule.default['dropdown-content'],
      style: {
        top: position.top,
        left: position.left
      },
      ref: dropdownRef,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: _DropdownFilterModule.default['dropdown-content__header'],
        children: labelOptions[filter?.type]
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: `${_DropdownFilterModule.default['dropdown-content__content']} ${_DropdownFilterModule.default[classes[type]]}`,
        ref: contentRef,
        children: [type === 'text' && /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
          type: "text",
          className: _DropdownFilterModule.default['input'],
          placeholder: VALUES.FILTER_COLUMN,
          value: filter?.selected,
          onChange: event => setSearchField(event.target.value)
        }), type === 'option' && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "pb-2",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: _DropdownFilterModule.default['find-option'],
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
              type: "text",
              className: _DropdownFilterModule.default['input'],
              placeholder: VALUES.FIND,
              onChange: event => searchForOptions(event.target.value)
            })
          }), options?.map((option, index) => /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
            className: _DropdownFilterModule.default['content__options'],
            onClick: () => handleCheckboxSelect(index),
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
              className: `${_DropdownFilterModule.default['option__checkbox']} ${option?.checked ? _DropdownFilterModule.default['option__checkbox--checked'] : ''}`,
              children: option?.checked ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdCheckBox, {}) : /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdCheckBoxOutlineBlank, {})
            }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
              children: option.name
            })]
          }, index))]
        }), type === 'between' && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "d-flex align-items-center gap-1 p-3",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Input.default, {
            type: "number",
            placeholder: VALUES.MIN,
            value: between?.min,
            min: 0,
            onChange: value => setBetween({
              ...between,
              min: value
            }),
            className: _DropdownFilterModule.default['between__input']
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_Input.default, {
            type: "number",
            placeholder: VALUES.MAX,
            value: between?.max,
            min: 0,
            onChange: value => setBetween({
              ...between,
              max: value
            }),
            className: _DropdownFilterModule.default['between__input']
          })]
        }), type === 'date' && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: `d-flex align-items-start gap-2`,
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: `${_DropdownFilterModule.default['calendar-divider']} w-45`,
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Calendar.default, {
              onChange: value => setDates({
                ...dates,
                start: value
              })
            })
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
            className: "w-45",
            children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_Calendar.default, {
              onChange: value => setDates({
                ...dates,
                end: value
              })
            })
          })]
        })]
      }), /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: _DropdownFilterModule.default['dropdown-content__footer'],
        children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: "d-flex gap-1",
          children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            className: _DropdownFilterModule.default['button__clear'],
            onClick: handleClear,
            children: VALUES.CLEAR
          }), type === 'option' && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            className: `${_DropdownFilterModule.default['button__select__all']} ${selected?.length === options?.length ? _DropdownFilterModule.default['button__select__all--disabled'] : ''}`,
            onClick: () => handleSelectAll(),
            children: VALUES.SELECT_ALL
          })]
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          className: "d-flex",
          children: /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            className: _DropdownFilterModule.default['button__apply'],
            onClick: applyFilter,
            children: VALUES.APPLY
          })
        })]
      })]
    })
  });
}
