// react-view-files/dist/components/MediaView/index.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaView = void 0;
var _react = _interopRequireWildcard(require("react"));
var _ContentViewer = _interopRequireDefault(require("../ContentViewer"));
var _MediaViewDocx = _interopRequireDefault(require("../MediaViewDocx"));
var _MediaViewPDF = _interopRequireDefault(require("../MediaViewPDF"));
var _MediaViewImage = _interopRequireDefault(require("../MediaViewImage"));
var _MediaViewVideo = _interopRequireDefault(require("../MediaViewVideo"));
var _MediaViewAudio = _interopRequireDefault(require("../MediaViewAudio"));
var _MediaViewTXT = _interopRequireDefault(require("../MediaViewTXT"));
var _MediaPreview = require("../MediaPreview");
var _extensions = require("../../extensions");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
var MediaView = exports.MediaView = function MediaView(_ref) {
  var file = _ref.file,
    _onRename = _ref.onRename,
    _onDelete = _ref.onDelete;
  var _useState = (0, _react.useState)(null),
    _useState2 = _slicedToArray(_useState, 2),
    selectedFile = _useState2[0],
    setSelectedFile = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    fileExtension = _useState4[0],
    setFileExtension = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = _slicedToArray(_useState5, 2),
    visible = _useState6[0],
    setVisible = _useState6[1];
  var _useState7 = (0, _react.useState)(""),
    _useState8 = _slicedToArray(_useState7, 2),
    childrenLink = _useState8[0],
    setChildrenLink = _useState8[1];
  var word = _extensions.extensionGroup.word,
    pdf = _extensions.extensionGroup.pdf,
    image = _extensions.extensionGroup.image,
    video = _extensions.extensionGroup.video,
    audio = _extensions.extensionGroup.audio,
    txt = _extensions.extensionGroup.txt;
  (0, _react.useEffect)(function () {
    setSelectedFile(file);
  }, [file]);
  var closedContent = function closedContent() {
    setVisible(false);
    if (childrenLink) URL.revokeObjectURL(childrenLink);
  };
  var handleView = function handleView(value, extension) {
    setSelectedFile(value);
    setFileExtension(extension);
    setVisible(true);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    id: "media-view-container",
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_MediaPreview.MediaPreview, {
      file: selectedFile,
      onClick: function onClick(value, extension) {
        return handleView(value, extension);
      },
      onRename: function onRename(value) {
        return _onRename === null || _onRename === void 0 ? void 0 : _onRename(value);
      },
      onDelete: function onDelete(value) {
        return _onDelete === null || _onDelete === void 0 ? void 0 : _onDelete(value);
      }
    }), visible && /*#__PURE__*/(0, _jsxRuntime.jsxs)(_ContentViewer["default"], {
      file: selectedFile,
      type: fileExtension,
      visible: visible,
      onClose: function onClose() {
        return closedContent();
      },
      children: [word.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type) && /*#__PURE__*/(0, _jsxRuntime.jsx)(_MediaViewDocx["default"], {
        file: selectedFile
      }), pdf.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type) && /*#__PURE__*/(0, _jsxRuntime.jsx)(_MediaViewPDF["default"], {
        file: selectedFile
      }), image.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type) && /*#__PURE__*/(0, _jsxRuntime.jsx)(_MediaViewImage["default"], {
        file: selectedFile,
        onLink: function onLink(value) {
          return setChildrenLink(value);
        }
      }), video.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type) && /*#__PURE__*/(0, _jsxRuntime.jsx)(_MediaViewVideo["default"], {
        file: selectedFile,
        onLink: function onLink(value) {
          return setChildrenLink(value);
        }
      }), audio.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type) && /*#__PURE__*/(0, _jsxRuntime.jsx)(_MediaViewAudio["default"], {
        file: selectedFile,
        onLink: function onLink(value) {
          return setChildrenLink(value);
        }
      }), txt.includes(selectedFile === null || selectedFile === void 0 ? void 0 : selectedFile.type) && /*#__PURE__*/(0, _jsxRuntime.jsx)(_MediaViewTXT["default"], {
        file: selectedFile,
        onLink: function onLink(value) {
          return setChildrenLink(value);
        }
      })]
    })]
  });
};
