// blend-editor/dist/components/BlendEditor/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BlendEditor;
var _react = _interopRequireWildcard(require("react"));
var _useOut = _interopRequireDefault(require("../../useOut"));
var _Editor = _interopRequireDefault(require("../Editor"));
require("./BlendEditor.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function BlendEditor(_ref) {
  let {
    content,
    placeholder,
    onSaved,
    focused,
    readOnly,
    maxHeight,
    useBorder,
    onBlur,
    onChange
  } = _ref;
  const [editing, setEditing] = (0, _react.useState)(false);
  const [onFocus, setOnFocus] = (0, _react.useState)(false);
  const [newContent, setNewContent] = (0, _react.useState)("");
  const [value, setValue] = (0, _react.useState)("");
  const editorRef = (0, _react.useRef)(null);
  (0, _react.useEffect)(() => {
    if (content) {
      setNewContent(content);
      setValue(content);
    }
  }, [content]);
  (0, _react.useEffect)(() => {
    const handleKeyDown = event => {
      if (event.keyCode === 27) {
        setEditing(false);
        if ((newContent === null || newContent === void 0 ? void 0 : newContent.length) === (content === null || content === void 0 ? void 0 : content.length)) setNewContent(content || "");
      }
    };
    if (editing) {
      document.addEventListener("keydown", handleKeyDown);
    }
    return () => {
      document.removeEventListener("keydown", handleKeyDown);
    };
  }, [editing]);
  (0, _react.useEffect)(() => {
    if (focused) {
      setOnFocus(focused);
      setEditing(true);
    }
  }, [focused]);
  (0, _useOut.default)(editorRef, () => onBlur === null || onBlur === void 0 ? void 0 : onBlur({
    content: {
      html: newContent
    }
  }));
  const handleEditing = () => {
    if (readOnly) return;
    setEditing(true);
    setOnFocus(true);
  };
  const handleSave = () => {
    setOnFocus(false);
    setEditing(false);
    onSaved({
      content: {
        html: newContent
      }
    });
    setValue(newContent);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "blend-editor-container",
    ref: editorRef,
    children: !editing ? /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "blend-editor-container__reading ".concat(readOnly && "blend-editor-container__reading--read-only", " ").concat(useBorder && "blend-editor-container__reading--border"),
      onClick: handleEditing,
      style: {
        ...(maxHeight && {
          maxHeight: maxHeight
        })
      },
      children: content ? /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        dangerouslySetInnerHTML: {
          __html: newContent
        }
      }) : placeholder
    }) : /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "blend-editor-container__editor",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_Editor.default, {
        content: value,
        focused: onFocus,
        onChange: _ref2 => {
          let {
            value
          } = _ref2;
          setNewContent(value), onChange === null || onChange === void 0 ? void 0 : onChange({
            content: {
              html: value
            }
          });
        }
      }), onSaved && actions && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: "blend-editor-container__editor__actions",
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
          className: "blend-editor-container__editor__actions--save",
          onClick: handleSave,
          children: "Salvar"
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("button", {
          className: "blend-editor-container__editor__actions--cancel",
          onClick: () => setEditing(false),
          children: "Cancelar"
        })]
      })]
    })
  });
}
