// powerful-react-table/dist/components/DatePicker/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DatePicker;
var _react = _interopRequireWildcard(require("react"));
var _useOut = _interopRequireDefault(require("../../useOut"));
var _md = require("react-icons/md");
var _Calendar = _interopRequireDefault(require("../Calendar"));
require("./DatePicker.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function DatePicker(_ref) {
  let {
    value,
    label,
    placeholder,
    minDate,
    maxDate,
    onChange
  } = _ref;
  const [state, setState] = (0, _react.useState)({
    date: undefined,
    inputDate: '',
    displayCalendar: false
  });
  const containerRef = (0, _react.useRef)(null);
  const contentRef = (0, _react.useRef)(null);
  const inputRef = (0, _react.useRef)(null);
  (0, _react.useEffect)(() => {
    setState({
      ...state,
      date: value,
      inputDate: value !== undefined ? value : ''
    });
  }, [value]);
  (0, _useOut.default)(containerRef, () => {
    setState({
      ...state,
      displayCalendar: false
    });
  });
  const handleToggleDropdown = event => {
    if (state?.displayCalendar) {
      return setState({
        ...state,
        displayCalendar: false
      });
    }
    calculatePosition(event);
    setState({
      ...state,
      displayCalendar: true
    });
  };
  const calculatePosition = event => {
    if (event) {
      const parentRect = event?.target?.getBoundingClientRect();
      const spaceAbove = parentRect.top;
      const spaceBelow = window.innerHeight - parentRect.bottom;
      setTimeout(() => {
        if (spaceAbove < spaceBelow) {
          contentRef.current.classList.remove('prt-date-picker-container__selector--dropup');
        } else {
          contentRef.current.classList.add('prt-date-picker-container__selector--dropup');
        }
      }, 1);
    }
  };
  const onBlur = () => {
    setState({
      ...state,
      date: state?.inputDate
    });
    onChange?.(state?.inputDat);
  };
  const onSelected = value => {
    setState({
      ...state,
      date: value,
      inputDate: value,
      displayCalendar: false
    });
    onChange?.(value);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "prt-date-picker-container",
    ref: containerRef,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "prt-date-picker-container__content",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
        className: `prt-date-picker-container__content--input ${state?.displayCalendar && 'prt-date-picker-container__content--input--no-focus'}`,
        type: "text",
        value: state?.inputDate,
        onBlur: onBlur,
        ref: inputRef,
        readOnly: true
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineCalendarToday, {
        className: "prt-date-picker-container__content--icon",
        onClick: event => handleToggleDropdown(event),
        onMouseOver: () => inputRef.current.style.borderColor = '#3772ff'
      })]
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: `prt-date-picker-container__selector ${state?.displayCalendar && 'prt-date-picker-container__selector--visible'}`,
      ref: contentRef,
      children: state?.displayCalendar && /*#__PURE__*/(0, _jsxRuntime.jsx)(_Calendar.default, {
        className: "prt-date-picker-container__selector--custom",
        selected: state?.date,
        minDate: minDate,
        maxDate: maxDate,
        onChange: value => onSelected(value)
      })
    })]
  });
}
