// powerful-react-table/dist/components/FilteredItems/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = FilteredItems;
var _react = _interopRequireDefault(require("react"));
var _useTable = require("../useTable");
var _md = require("react-icons/md");
var _fi = require("react-icons/fi");
var _localeConfig = require("../../localeConfig");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function FilteredItems() {
  const {
    searchFilter,
    setSearchFilter,
    setRemoveFilter,
    tableId
  } = (0, _useTable.useTable)();
  const localeText = (0, _localeConfig.getLocaleText)();
  const removeFilter = index => {
    const filters = searchFilter;
    const removed = filters.splice(index, 1);
    const storage = localStorage.getItem('__magnunTablePreferences');
    const savedFilter = JSON.parse(storage)?.filters;
    delete savedFilter[tableId][removed[0]?.label?.toUpperCase()];
    localStorage.setItem('__magnunTablePreferences', JSON.stringify({
      ...JSON.parse(storage),
      filters: savedFilter
    }));
    setSearchFilter([...filters]);
    setRemoveFilter(true);
  };
  const getValues = filter => {
    if (filter?.type === 'option') {
      return filter?.value?.join(', ');
    }
    if (filter?.type === 'between') {
      return `${filter?.value?.min} - ${filter?.value?.max}`;
    }
    if (filter?.type === 'date') {
      return `${filter?.value?.start} - ${filter?.value?.end}`;
    }
    return filter?.value || '';
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_jsxRuntime.Fragment, {
    children: searchFilter?.length ? /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: 'prt-filtered-items',
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
        className: 'prt-filtered-items__legend',
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_fi.FiFilter, {}), " ", localeText.toolbarFilterWithColon]
      }), searchFilter?.map((filter, index) => /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        children: filter?.label && /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
          className: 'prt-filtered-items__item',
          children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
            style: {
              color: '#808097'
            },
            children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
              style: {
                fontWeight: 500,
                color: '#162929'
              },
              children: filter.label
            }), ": ", getValues(filter)]
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdOutlineClose, {
            onClick: () => removeFilter(index)
          })]
        }, index)
      }, index))]
    }) : /*#__PURE__*/(0, _jsxRuntime.jsx)(_jsxRuntime.Fragment, {})
  });
}
