// powerful-react-table/dist/components/Export/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Export;
var _react = _interopRequireDefault(require("react"));
var _moment = _interopRequireDefault(require("moment"));
var XLSX = _interopRequireWildcard(require("xlsx"));
var _useTable = require("../useTable");
var _DropdownMenu = _interopRequireDefault(require("../DropdownMenu"));
var _fi = require("react-icons/fi");
var _ExportModule = _interopRequireDefault(require("./Export.module.scss"));
var _localeConfig = require("../../localeConfig");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function Export(_ref) {
  let {
    data = [],
    headers = [],
    title = '',
    tableRef
  } = _ref;
  const {
    selected
  } = (0, _useTable.useTable)();
  const locateText = (0, _localeConfig.getLocaleText)();
  const fileName = `${title?.toLowerCase()?.replaceAll(' ', '_')}_${(0, _moment.default)().format('YYYYMMDDHHmmss')}`;
  const exportToCSV = () => {
    const filteredData = data.filter(item => selected.includes(Number(item.getAttribute('tr-key'))));
    let csv = headers.map(header => header.name)?.join(',') + '\n';
    csv += filteredData.map(child => Array.from(child.childNodes).map(td => td.innerText.replace(/"/g, '""')).map(tdText => tdText.includes(',') ? `"${tdText}"` : tdText).join(',')).join('\n');
    const link = document.createElement('a');
    link.href = 'data:text/csvcharset=utf-8,' + encodeURIComponent(csv);
    link.download = `${fileName}.csv`;
    link.click();
  };
  const exportToXLSX = () => {
    const filteredData = data.filter(item => selected.includes(Number(item.getAttribute('tr-key'))));
    const wb = XLSX.utils.book_new();
    const ws = XLSX.utils.aoa_to_sheet([headers.map(header => header.name), ...filteredData.map(child => Array.from(child.childNodes).map(td => td.innerText))]);
    XLSX.utils.book_append_sheet(wb, ws, 'Sheet1');
    XLSX.writeFile(wb, `${fileName}.xlsx`);
  };
  const handleDownload = value => {
    if (value?.type === 'xlsx') exportToXLSX();
    if (value?.type === 'csv') exportToCSV();
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: _ExportModule.default['prt-export-container'],
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)(_DropdownMenu.default, {
      options: [{
        name: locateText.toolbarExportExcel,
        type: 'xlsx'
      }, {
        name: locateText.toolbarExportCSV,
        type: 'csv'
      }],
      trigger: /*#__PURE__*/(0, _jsxRuntime.jsxs)("button", {
        style: {
          border: 'none',
          backgroundColor: '#005FED',
          height: '34px',
          borderRadius: '4px',
          display: 'flex',
          alignItems: 'center',
          gap: '6px',
          color: '#fff',
          padding: '0 16px',
          cursor: 'pointer'
        },
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(_fi.FiDownload, {
          style: {
            fontSize: '16px'
          }
        }), " ", locateText.toolbarExport]
      }),
      customTriggerClassName: _ExportModule.default['export-table-items-button'],
      onClicked: value => handleDownload(value)
    })
  });
}
