// powerful-react-table/dist/components/Input/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Input;
var _react = _interopRequireWildcard(require("react"));
require("./Input.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function Input(_ref) {
  let {
    type = '',
    placeholder = '',
    value,
    label = '',
    className = '',
    onChange,
    money,
    focused = false,
    min = 0,
    max = 0,
    styles,
    iconBefore
  } = _ref;
  const inputRef = (0, _react.useRef)(null);
  const [input, setInput] = (0, _react.useState)('');
  (0, _react.useEffect)(() => {
    if (value) {
      setInput(value);
    }
  }, [value]);
  (0, _react.useEffect)(() => {
    if (focused) {
      setTimeout(() => {
        inputRef?.current?.focus();
      }, 0);
    }
  }, [focused]);
  const changeInput = event => {
    setInput(event.target.value);
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: 'prt-input-container',
    children: [iconBefore && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      className: "prt-input-container__icon-before",
      children: iconBefore
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
      ref: inputRef,
      type: type,
      value: input,
      onChange: event => {
        onChange?.(event.target.value);
        changeInput(event);
      },
      className: `input ${iconBefore && 'input__has-icon-before'} ${className}`,
      placeholder: placeholder,
      min: min,
      max: max,
      style: styles
    })]
  });
}
