// powerful-react-table/dist/components/DropdownMenu/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DropdownMenu;
var _react = _interopRequireWildcard(require("react"));
var _useOut = _interopRequireDefault(require("../../useOut"));
var _md = require("react-icons/md");
require("./DropdownMenu.scss");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function DropdownMenu(_ref) {
  let {
    options,
    onClicked,
    direction,
    icon,
    dropdownVisible,
    trigger,
    customTriggerClassName,
    disabled = false
  } = _ref;
  const [dropdownOpen, setDropdownOpen] = (0, _react.useState)(false);
  const [position, setPosition] = (0, _react.useState)({
    top: 0,
    left: 0
  });
  const dropdownRef = (0, _react.useRef)(null);
  const dropdownContentRef = (0, _react.useRef)(null);
  const childRef = (0, _react.useRef)(null);
  (0, _useOut.default)(dropdownRef, () => {
    setDropdownOpen(false);
    dropdownVisible?.(false);
  });
  const handleToggleDropdown = event => {
    if (dropdownOpen) {
      setDropdownOpen(false);
      dropdownVisible?.(false);
      return;
    }
    calculatePosition(event);
    setDropdownOpen(true);
    dropdownVisible?.(true);
  };
  function navigateTo(url, onNavigateCallback) {
    try {
      window.location.href = url;
      if (onNavigateCallback) {
        onNavigateCallback();
      }
    } catch (error) {
      console.error("Error during navigation:", error);
    }
  }
  const handleNavigateItem = function (url) {
    let redirect = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
    if (trigger) return;
    setDropdownOpen(false);
    dropdownVisible?.(false);
    if (!url) return;
    if (redirect) navigateTo(url);
  };
  const calculatePosition = event => {
    if (dropdownRef.current && event) {
      const parentRect = event?.target?.getBoundingClientRect();
      const dropdownRect = dropdownContentRef.current.getBoundingClientRect();
      const spaceRight = window.innerWidth - parentRect.right;
      const spaceAbove = parentRect.top;
      const spaceBelow = window.innerHeight - parentRect.bottom;
      let newPosition = {
        left: spaceRight >= dropdownRect.width ? parentRect.left : parentRect.right - (dropdownRect?.width || 160)
      };
      setTimeout(() => {
        const childHeight = childRef.current.getBoundingClientRect()?.height + 18;
        if (spaceAbove < spaceBelow) {
          newPosition.top = parentRect.bottom + 10;
        } else {
          newPosition.top = parentRect.top - childHeight;
        }
        setPosition(newPosition);
      }, 1);
    }
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    className: "dmc-dropdown-menu-container",
    ref: dropdownRef,
    children: [!trigger ? /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: 'dropdown__header',
      onClick: event => handleToggleDropdown(event),
      children: icon ? /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
        className: 'dmc-dropdown-icon',
        children: icon
      }) : /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdMoreVert, {
        className: 'dmc-dropdown-icon'
      })
    }) : /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      onClick: event => handleToggleDropdown(event),
      className: `dmc-trigger-dropdown ${customTriggerClassName}`,
      children: trigger
    }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: `dropdown__content ${dropdownOpen && 'visible'} ${direction === 'top' && 'reverse'} ${trigger && 'dropdown__content--large'} ${disabled && 'dropdown__content--disabled'}`,
      ref: dropdownContentRef,
      style: {
        top: position.top,
        left: position.left
      },
      children: /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {
        ref: childRef,
        children: options?.map((option, index) => /*#__PURE__*/(0, _jsxRuntime.jsxs)("li", {
          className: "dropdow__item",
          onClick: () => {
            handleNavigateItem(option?.url, option?.redirect);
            onClicked?.(option);
          },
          children: [option.icon && /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            className: "dropdow__item--icon",
            children: option.icon
          }), /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
            children: option.name
          })]
        }, index))
      })
    })]
  });
}
