// magnun/src/mocks/messages.js
import messageAudio1 from '../assets/files/audios/WhatsApp Ptt 2024-08-26 at 08.35.18.ogg'
import messageAudio2 from '../assets/files/audios/WhatsApp Ptt 2024-08-26 at 08.35.08.ogg'
import messageVideo1 from '../assets/files/videos/Elofy - Google Chrome 2024-07-08 11-35-07.mp4'

const messages = [
  {
    id: 1,
    uuid: 'b3b7f2a8-14f0-4a5d-b929-cb5a9d8c2f5a',
    senderId: 220,
    receiverId: 2,
    content: 'Hey John, how are you doing?',
    createdAt: '2024-08-14T10:15:00',
    read: false,
    deleted: false,
    type: 'received',
  },
  {
    id: 2,
    uuid: '9f6c8b1d-2c3d-4e5f-6a7b-8c9d0e1f2g3h',
    senderId: 220,
    receiverId: 2,
    content: 'Just wanted to let you know about the meeting tomorrow.',
    createdAt: '2024-08-14T10:30:00',
    read: false,
    deleted: false,
    type: 'received',
  },
  {
    id: 3,
    uuid: 'dc8e0bce-8c21-4e5a-b7f7-d7e5d3a1b7b9',
    senderId: 220,
    receiverId: 2,
    content:
      'I’m doing great! Just working on a new project. Would you like to catch up over coffee?',
    createdAt: '2024-08-14T10:20:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 4,
    uuid: 'a7b8c9d0-e1f2-3g4h-5i6j-7k8l9m0n1o2p',
    senderId: 220,
    receiverId: 2,
    content: "I'll send over the documents by the end of the day.",
    createdAt: '2024-08-14T10:45:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 5,
    uuid: 'e3e1f0e6-16d8-4cb7-95ed-29045dcd3f45',
    senderId: 220,
    receiverId: 2,
    content: 'Sounds good to me! Let’s meet at 10 AM.',
    createdAt: '2024-08-14T10:30:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 6,
    uuid: 'b4c5d6e7-8f9g-0h1i-2j3k-4l5m6n7o8p9q',
    senderId: 220,
    receiverId: 3,
    content: 'Can you review the report and give me your feedback?',
    createdAt: '2024-08-14T11:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 7,
    uuid: 'c9d0e1f2-3g4h-5i6j-7k8l-9m0n1o2p3q4r',
    senderId: 220,
    receiverId: 3,
    content: "Don't forget about the project deadline next week.",
    createdAt: '2024-08-14T11:30:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 8,
    uuid: '2a2b3c4d-5e6f-7g8h-9i0j-k1lmnopqrs',
    senderId: 220,
    receiverId: 3,
    content: 'Sure, what time works for you?',
    createdAt: '2024-08-14T11:05:00',
    read: true,
    deleted: false,
    type: 'sent',
  },
  {
    id: 9,
    uuid: '3b4c5d6e-7f8g-9h0i-1j2k-3l4m5n6o7p8q',
    senderId: 220,
    receiverId: 4,
    content: 'Hey, just checking in to see how everything is going.',
    createdAt: '2024-08-14T12:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 10,
    uuid: '4c4d5e6f-7g8h-9i0j-1k2l-m3nopqrst',
    senderId: 220,
    receiverId: 3,
    content: '2 PM works for me. Talk to you then!',
    createdAt: '2024-08-14T11:15:00',
    read: true,
    deleted: false,
    type: 'sent',
  },
  {
    id: 11,
    uuid: '5d6e7f8g-9h0i-1j2k-3l4m-5n6o7p8q9r0s',
    senderId: 220,
    receiverId: 4,
    content: "I'll get back to you with an update later today.",
    createdAt: '2024-08-14T13:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 12,
    uuid: '6e7f8g9h-0i1j-2k3l-4m5n-6o7p8q9r0s1t',
    senderId: 220,
    receiverId: 5,
    content: 'Please confirm if you received my earlier message.',
    createdAt: '2024-08-14T14:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 13,
    uuid: '7f8g9h0i-1j2k-3l4m-5n6o-7p8q9r0s1t2u',
    senderId: 220,
    receiverId: 6,
    content: 'Looking forward to your response about the new proposal.',
    createdAt: '2024-08-14T15:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 14,
    uuid: '8g9h0i1j-2k3l-4m5n-6o7p-8q9r0s1t2u3v',
    senderId: 220,
    receiverId: 2,
    content: 'I have sent you the updated files. Please check.',
    createdAt: '2024-08-14T16:00:00',
    read: false,
    deleted: false,
    type: 'received',
  },
  {
    id: 14,
    uuid: '8g9h0i1j-2k3l-4m5n-6o7p-8q9r0s1t2su',
    senderId: 220,
    receiverId: 2,
    content: messageAudio1,
    createdAt: '2024-08-14T16:01:00',
    read: false,
    deleted: false,
    type: 'received',
  },
  {
    id: 15,
    uuid: '9h0i1j2k-3l4m-5n6o-7p8q-9r0s1t2u3v4w',
    senderId: 220,
    receiverId: 4,
    content:
      'Peter, I saw your request for the documents. I’ll send them over soon.',
    createdAt: '2024-08-14T14:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 16,
    uuid: '0i1j2k3l-4m5n-6o7p-8q9r-0s1t2u3v4w5x',
    senderId: 220,
    receiverId: 5,
    content: 'Reminder about the meeting tomorrow.',
    createdAt: '2024-08-14T17:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 17,
    uuid: '1j2k3l4m-5n6o-7p8q-9r0s-1t2u3v4w5x6y',
    senderId: 220,
    receiverId: 3,
    content: 'Anna, could you please update me on the latest project status?',
    createdAt: '2024-08-14T15:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 18,
    uuid: '2k3l4m5n-6o7p-8q9r-0s1t-2u3v4w5x6y7z',
    senderId: 220,
    receiverId: 6,
    content: 'When can we schedule a meeting to discuss the new features?',
    createdAt: '2024-08-14T18:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 19,
    uuid: '3l4m5n6o-7p8q-9r0s-1t2u-3v4w5x6y7z8a',
    senderId: 220,
    receiverId: 2,
    content: 'I’m preparing the presentation for our meeting tomorrow.',
    createdAt: '2024-08-14T19:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 20,
    uuid: '4m5n6o7p-8q9r-0s1t-2u3v-4w5x6y7z8a9b',
    senderId: 220,
    receiverId: 3,
    content: 'Please review the attached file and provide your feedback.',
    createdAt: '2024-08-14T20:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 21,
    uuid: '5n6o7p8q-9r0s-1t2u-3v4w-5x6y7z8a9b0c',
    senderId: 220,
    receiverId: 2,
    content: 'Smith, did you get a chance to review the report I sent you?',
    createdAt: '2024-08-14T17:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 40,
    uuid: '5n6o7p8q-9r0s-1t2u-3v4w-5x6y7z8a9b0c',
    senderId: 220,
    receiverId: 2,
    content:
      'Yes, I was able to review it. Below is a video showing evidence of the review.',
    createdAt: '2024-08-14T17:00:00',
    read: false,
    deleted: false,
    type: 'received',
  },
  {
    id: 42,
    uuid: '5n6o7p8q-9r0s-1t2u-3v4w-5x6y7z8a9b0c',
    senderId: 220,
    receiverId: 2,
    content: messageVideo1,
    createdAt: '2024-08-14T17:00:00',
    read: false,
    deleted: false,
    type: 'received',
  },
  {
    id: 22,
    uuid: '6o7p8q9r-0s1t-2u3v-4w5x-6y7z8a9b0c1d',
    senderId: 220,
    receiverId: 4,
    content: "I'll follow up with the team and update you shortly.",
    createdAt: '2024-08-14T21:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 23,
    uuid: '7p8q9r0s-1t2u-3v4w-5x6y-7z8a9b0c1d2e',
    senderId: 220,
    receiverId: 3,
    content: 'Anna, can we reschedule our chat to the morning?',
    createdAt: '2024-08-14T18:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 24,
    uuid: '8q9r0s1t-2u3v-4w5x-6y7z-8a9b0c1d2e3f',
    senderId: 220,
    receiverId: 2,
    content: 'Please confirm receipt of this message.',
    createdAt: '2024-08-14T22:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 47,
    uuid: '8q9r0s1t-2u3v-4w5x-6y7z-8a9b0c1d2e3f',
    senderId: 220,
    receiverId: 2,
    content: messageAudio2,
    createdAt: '2024-08-14T22:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 25,
    uuid: '9r0s1t2u-3v4w-5x6y-7z8a-9b0c1d2e3f4g',
    senderId: 220,
    receiverId: 4,
    content: 'The documents you requested have been sent.',
    createdAt: '2024-08-14T23:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 26,
    uuid: '0s1t2u3v-4w5x-6y7z-8a9b-0c1d2e3f4g5h',
    senderId: 220,
    receiverId: 5,
    content: 'Reminder: Please review the attached file and provide feedback.',
    createdAt: '2024-08-14T23:30:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 27,
    uuid: '1t2u3v4w-5x6y-7z8a-9b0c-1d2e3f4g5h6i',
    senderId: 220,
    receiverId: 6,
    content:
      'Please review the updated proposal and let me know your thoughts.',
    createdAt: '2024-08-15T00:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 28,
    uuid: '2u3v4w5x-6y7z-8a9b-0c1d-2e3f4g5h6i7j',
    senderId: 220,
    receiverId: 3,
    content: 'Sure, I’ll take a look at them this evening.',
    createdAt: '2024-08-14T20:05:00',
    read: true,
    deleted: false,
    type: 'sent',
  },
  {
    id: 29,
    uuid: '3v4w5x6y-7z8a-9b0c-1d2e-3f4g5h6i7j8k',
    senderId: 220,
    receiverId: 5,
    content: 'Looking forward to your feedback on the new project.',
    createdAt: '2024-08-15T01:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 30,
    uuid: '4w5x6y7z-8a9b-0c1d-2e3f-4g5h6i7j8k9l',
    senderId: 220,
    receiverId: 6,
    content: 'Can we schedule a quick call to discuss the changes?',
    createdAt: '2024-08-15T02:00:00',
    read: false,
    deleted: false,
    type: 'sent',
  },
  {
    id: 30,
    uuid: '3w5x6y7z-8a9b-0n1d-2e3f-4g5h6i7j8k9l',
    senderId: 220,
    receiverId: 1,
    content: "What's on your project?",
    createdAt: '2024-08-17 16:28:42',
    read: false,
    deleted: false,
    type: 'sent',
  },
]

export default messages
