// blend-editor/dist/components/Dropdown/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DropdownTitle = exports.DropdownItem = exports.DropdownContent = exports.Dropdown = void 0;
var _react = _interopRequireWildcard(require("react"));
var _uuid = require("uuid");
var _useOut = _interopRequireDefault(require("../../useOut"));
var _useDropdown = require("./useDropdown");
var _dropdown = _interopRequireDefault(require("./dropdown"));
var _md = require("react-icons/md");
var _DropdownModule = _interopRequireDefault(require("./Dropdown.module.scss"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const Dropdown = _ref => {
  let {
    children,
    placement,
    trigger,
    triggerClassName,
    hideDropdownIcon
  } = _ref;
  const parentRef = (0, _react.useRef)((0, _uuid.v4)());
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "".concat(_DropdownModule.default['dropdown-container']),
    ref: parentRef,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_dropdown.default, {
      placement: placement,
      trigger: trigger,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(DropdownTitle, {
        className: triggerClassName,
        hideIcon: hideDropdownIcon
      }), children]
    })
  });
};
exports.Dropdown = Dropdown;
const DropdownTitle = _ref2 => {
  var _trigger$type;
  let {
    className,
    hideIcon
  } = _ref2;
  const {
    visible,
    setVisible,
    trigger
  } = (0, _useDropdown.useDropdown)();
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("span", {
    className: "d-flex align-items-center gap-1 ".concat(typeof trigger === 'string' || trigger !== null && trigger !== void 0 && (_trigger$type = trigger.type) !== null && _trigger$type !== void 0 && _trigger$type.hasOwnProperty('prototype') ? _DropdownModule.default['dropdown-container__trigger'] : '', " ").concat(className || ''),
    onClick: () => setVisible(!visible),
    children: [trigger, !hideIcon && /*#__PURE__*/(0, _jsxRuntime.jsx)(_md.MdKeyboardArrowDown, {})]
  });
};
exports.DropdownTitle = DropdownTitle;
const DropdownContent = _ref3 => {
  let {
    className,
    children
  } = _ref3;
  const {
    visible,
    setVisible,
    placement
  } = (0, _useDropdown.useDropdown)();
  const dropdownRef = (0, _react.useRef)((0, _uuid.v4)());
  const [parentRef, setParentRef] = (0, _react.useState)({
    current: null
  });
  const [calculatedDirection, setCalculatedDirection] = (0, _react.useState)(placement);
  (0, _react.useEffect)(() => {
    setTimeout(() => {
      var _dropdownRef$current;
      setParentRef({
        current: dropdownRef === null || dropdownRef === void 0 || (_dropdownRef$current = dropdownRef.current) === null || _dropdownRef$current === void 0 ? void 0 : _dropdownRef$current.parentElement
      });
    }, 1);
  }, []);
  (0, _react.useEffect)(() => {
    var _document$getElementB;
    const recalculatePosition = () => {
      const parentElement = parentRef.current;
      const dropdownElement = dropdownRef.current;
      if (!parentElement || !dropdownElement) return;
      const parentRect = parentElement.getBoundingClientRect();
      const dropdownRect = dropdownElement.getBoundingClientRect();

      // Considera a altura da barra de rolagem
      const scrollBarHeight = window.innerWidth - document.documentElement.clientWidth;
      const spaceBelow = window.innerHeight - parentRect.bottom - scrollBarHeight;
      const spaceAbove = parentRect.top - scrollBarHeight;
      let newDirection = placement;
      if (!placement) {
        newDirection = spaceAbove < dropdownRect.height ? 'default-top' : 'default-bottom';
      }
      if (placement === 'bottom-end') {
        newDirection = spaceAbove < dropdownRect.height ? 'top-end' : 'bottom-end';
      }
      if (placement === 'top-end') {
        newDirection = spaceBelow < dropdownRect.height ? 'bottom-end' : 'top-end';
      }
      if (placement === 'bottom-start' && spaceBelow < dropdownRect.height) {
        newDirection = spaceAbove < dropdownRect.height ? 'top-start' : 'bottom-start';
      }
      if (placement === 'top-start' && spaceAbove < dropdownRect.height) {
        newDirection = spaceBelow < dropdownRect.height ? 'bottom-start' : 'top-start';
      }
      setCalculatedDirection(newDirection);
    };
    const handleScroll = () => {
      recalculatePosition();
    };
    recalculatePosition();
    window.addEventListener('resize', recalculatePosition);
    (_document$getElementB = document.getElementById('default-main')) === null || _document$getElementB === void 0 || _document$getElementB.addEventListener('scroll', handleScroll);
    return () => {
      var _document$getElementB2;
      window.removeEventListener('resize', recalculatePosition);
      (_document$getElementB2 = document.getElementById('default-main')) === null || _document$getElementB2 === void 0 || _document$getElementB2.removeEventListener('scroll', handleScroll);
    };
  }, [placement, parentRef]);
  (0, _useOut.default)(parentRef, () => setVisible(false));
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "".concat(_DropdownModule.default['dropdown-container__content'], " ").concat(className, " ").concat(visible ? "".concat(_DropdownModule.default.visible) : '', " ").concat(_DropdownModule.default[calculatedDirection]),
    ref: dropdownRef,
    children: /*#__PURE__*/(0, _jsxRuntime.jsx)("ul", {
      className: "".concat(_DropdownModule.default.content__list),
      children: children
    })
  });
};
exports.DropdownContent = DropdownContent;
const DropdownItem = _ref4 => {
  let {
    children,
    onClick
  } = _ref4;
  const {
    setVisible
  } = (0, _useDropdown.useDropdown)();
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: "".concat(_DropdownModule.default.dropdown__item),
    onClick: event => {
      onClick === null || onClick === void 0 || onClick(event);
      setVisible(false);
    },
    children: children
  });
};
exports.DropdownItem = DropdownItem;
