// powerful-react-table/dist/components/Checkbox/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Checkbox;
var _react = _interopRequireWildcard(require("react"));
var _uuid = require("uuid");
var _CheckboxModule = _interopRequireDefault(require("./Checkbox.module.scss"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function Checkbox(_ref) {
  let {
    value = '',
    label = '',
    isChecked = false,
    onChange,
    isIndeterminate = false
  } = _ref;
  const [checked, setChecked] = (0, _react.useState)(false);
  const inputRef = (0, _react.useRef)(null);
  const checkboxId = `checkbox-${(0, _uuid.v4)()}`;
  (0, _react.useEffect)(() => {
    setChecked(isChecked);
  }, [isChecked]);
  const handleChange = event => {
    setChecked(event.target.checked);
    onChange?.({
      value: value,
      checked: event.target.checked,
      event
    });
  };
  return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    className: `${_CheckboxModule.default['prt-checkbox-container']} ${!label && _CheckboxModule.default['prt-checkbox-container__only-input']}`,
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
      className: "d-flex align-items-center gap-1",
      children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("input", {
        type: "checkbox",
        value: value,
        id: checkboxId,
        checked: checked,
        className: `${_CheckboxModule.default['prt-checkbox-container__checkbox']} ${isIndeterminate && _CheckboxModule.default['prt-checkbox-container__checkbox--indeterminate']}`,
        ref: inputRef,
        onChange: handleChange
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)("label", {
        className: _CheckboxModule.default['prt-checkbox-container__label'],
        htmlFor: checkboxId,
        children: label
      })]
    })
  });
}
