// powerful-react-table/dist/_utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.removeAccents = exports.insertChildrenToDOM = exports.findInnerText = void 0;
var _react = _interopRequireDefault(require("react"));
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const accentsMap = {
  a: 'á|à|ã|â|À|Á|Ã|Â',
  e: 'é|è|ê|É|È|Ê',
  i: 'í|ì|î|Í|Ì|Î',
  o: 'ó|ò|ô|õ|Ó|Ò|Ô|Õ',
  u: 'ú|ù|û|ü|Ú|Ù|Û|Ü',
  c: 'ç|Ç',
  n: 'ñ|Ñ'
};
const findInnerText = element => {
  if (typeof element === 'string') return element;
  if (Array.isArray(element)) {
    for (const el of element) {
      const result = findInnerText(el);
      if (result) return result;
    }
  }
  if (element?.props?.children) return findInnerText(element.props.children);
  return null;
};
exports.findInnerText = findInnerText;
const removeAccents = str => Object.keys(accentsMap).reduce((acc, cur) => acc.replace(new RegExp(accentsMap[cur], 'g'), cur), str);
exports.removeAccents = removeAccents;
const insertChildrenToDOM = children => {
  return _react.default.Children.map(children, (child, index) => {
    if (typeof child === 'string') {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
        className: "td-item--apply-ellipsis",
        children: child
      }, index);
    } else if (/*#__PURE__*/_react.default.isValidElement(child)) {
      let updatedChild = child;
      if (child?.props && child?.props?.children) {
        const updatedChildren = insertChildrenToDOM(child?.props?.children);
        updatedChild = /*#__PURE__*/_react.default.cloneElement(child, {
          ...child?.props,
          children: updatedChildren
        });
      }
      updatedChild = /*#__PURE__*/_react.default.cloneElement(updatedChild, {
        key: index,
        className: `${updatedChild?.props && updatedChild?.props?.className ? updatedChild?.props?.className + ' ' : ''}td-item--apply-ellipsis`
      });
      return updatedChild;
    } else {
      return child;
    }
  });
};
exports.insertChildrenToDOM = insertChildrenToDOM;
