// powerful-react-table/dist/localeTextConstants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_LOCALE_TEXT = void 0;
const DEFAULT_LOCALE_TEXT = exports.DEFAULT_LOCALE_TEXT = {
  // Root
  noRowsLabel: 'No rows',
  noResultsOverlayLabel: 'No results found.',
  // Columns selector toolbar button text
  toolbarColumns: 'Columns',
  // Filters toolbar button text
  toolbarFilter: 'Filters',
  toolbarFilterWithColon: 'Filters:',
  // Quick filter toolbar field
  toolbarQuickFilterPlaceholder: 'Search…',
  // Export selector toolbar button text
  toolbarExport: 'Export',
  toolbarExportLabel: 'Export',
  toolbarExportCSV: 'Download as CSV',
  toolbarExportPrint: 'Print',
  toolbarExportExcel: 'Download as Excel',
  // Filter panel text
  filterBoxApplyFilter: 'Apply filter',
  filterBoxClearFilter: 'Clear',
  // Header filter operators text
  headerFilterOperatorContains: 'Contains',
  headerFilterOperatorBetween: 'Between',
  headerFilterOperatorDate: 'Date',
  // Column menu text
  columnMenuFilterBy: 'Filter by',
  columnMenuAutosizeThisColumn: 'Autosize this column',
  columnMenuMoveColumnLeft: 'Move column left',
  columnMenuMoveColumnRight: 'Move column right',
  columnMenuHideColumn: 'Hide column',
  // Pagination
  rowsPerPageLabel: 'Displaying per page'
};
